// @sap-ux/eslint-plugin-fiori-tools/lib/language/annotations/source-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FioriAnnotationSourceCode = exports.visitorKeys = void 0;
const plugin_kit_1 = require("@eslint/plugin-kit");
const odata_annotation_core_1 = require("@sap-ux/odata-annotation-core");
const traversal_step_1 = require("./traversal-step");
/**
 * Visitor keys mapping annotation node types to their traversable child properties.
 * Used to navigate the annotation AST during traversal.
 */
exports.visitorKeys = {
    [odata_annotation_core_1.ANNOTATION_FILE_TYPE]: ['references', 'namespace', 'targets'],
    [odata_annotation_core_1.ELEMENT_TYPE]: ['attributes', 'content'],
    [odata_annotation_core_1.NAMESPACE_TYPE]: [],
    [odata_annotation_core_1.REFERENCE_TYPE]: [],
    [odata_annotation_core_1.TARGET_TYPE]: ['terms'],
    [odata_annotation_core_1.ATTRIBUTE_TYPE]: [],
    [odata_annotation_core_1.TEXT_TYPE]: []
};
/**
 * Annotation Source Code representation.
 */
class FioriAnnotationSourceCode extends plugin_kit_1.TextSourceCodeBase {
    projectContext;
    /**
     * The AST of the source code.
     */
    ast;
    /**
     * Creates an instance of FioriAnnotationSourceCode.
     *
     * @param root0 - Object containing text, ast, and projectContext.
     * @param root0.text - The text content of the file.
     * @param root0.ast - The Annotation AST.
     * @param root0.projectContext - The project context associated with the given file.
     */
    constructor({ text, ast, projectContext }) {
        super({ text, ast });
        this.ast = ast;
        this.projectContext = projectContext;
    }
    /**
     * Traverse the source code and return the steps that were taken.
     *
     * @returns The steps that were taken while traversing the source code.
     */
    traverse() {
        const steps = [];
        const visit = (node, parent) => {
            steps.push(new traversal_step_1.AnnotationVisitNodeStep({
                target: node,
                phase: traversal_step_1.STEP_PHASE.ENTER,
                args: [node, parent]
            }));
            for (const key of exports.visitorKeys[node.type] ?? []) {
                // @ts-expect-error 7053
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                const child = node[key];
                if (!child) {
                    return;
                }
                if (Array.isArray(child)) {
                    for (const grandchild of child) {
                        visit(grandchild, node);
                    }
                }
                else if (node.type === 'element' && key === 'attributes') {
                    for (const grandchild of Object.values(child)) {
                        visit(grandchild, node);
                    }
                }
                else {
                    visit(child, node);
                }
            }
            steps.push(new traversal_step_1.AnnotationVisitNodeStep({
                target: node,
                phase: traversal_step_1.STEP_PHASE.EXIT,
                args: [node, parent]
            }));
        };
        visit(this.ast);
        return steps;
    }
    /**
     * Get location of node or token.
     *
     * @param node - Node or token
     * @returns Location
     */
    getLoc(node) {
        if (!node.range) {
            throw new Error('Node has no range');
        }
        return {
            start: {
                line: node.range.start.line + 1,
                column: node.range.start.character + 1
            },
            end: {
                line: node.range.end.line + 1,
                column: node.range.end.character + 1
            }
        };
    }
}
exports.FioriAnnotationSourceCode = FioriAnnotationSourceCode;
