// @sap-ux/eslint-plugin-fiori-tools/lib/utils/helpers.js
"use strict";
/**
 * Common AST helper functions for ESLint rules
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isType = isType;
exports.isIdentifier = isIdentifier;
exports.isMember = isMember;
exports.isCall = isCall;
exports.isLiteral = isLiteral;
exports.isProperty = isProperty;
exports.isObject = isObject;
exports.isArray = isArray;
exports.contains = contains;
exports.isString = isString;
exports.startsWith = startsWith;
exports.containsString = containsString;
exports.isWindow = isWindow;
exports.createIsWindowObject = createIsWindowObject;
exports.createRememberWindow = createRememberWindow;
exports.createIsDocument = createIsDocument;
exports.createIsDocumentObject = createIsDocumentObject;
exports.createRememberDocument = createRememberDocument;
exports.buildCalleePath = buildCalleePath;
exports.getMemberAsString = getMemberAsString;
exports.createIsHistory = createIsHistory;
exports.createIsHistoryObject = createIsHistoryObject;
exports.createRememberHistory = createRememberHistory;
exports.createIsLocation = createIsLocation;
exports.createIsLocationObject = createIsLocationObject;
exports.createRememberLocation = createRememberLocation;
exports.getLiteralOrIdentifierName = getLiteralOrIdentifierName;
exports.getIdentifierPath = getIdentifierPath;
exports.resolveIdentifierPath = resolveIdentifierPath;
exports.rememberInterestingVariable = rememberInterestingVariable;
exports.checkInterestingPath = checkInterestingPath;
exports.createVariableDeclaratorProcessor = createVariableDeclaratorProcessor;
exports.hasUnderscore = hasUnderscore;
exports.isInteger = isInteger;
exports.endsWith = endsWith;
exports.isForbiddenObviousApi = isForbiddenObviousApi;
exports.createStorageRuleHelpers = createStorageRuleHelpers;
exports.createPropertyChecker = createPropertyChecker;
exports.createDocumentBasedRuleVisitors = createDocumentBasedRuleVisitors;
exports.findDeepestExistingPath = findDeepestExistingPath;
// ------------------------------------------------------------------------------
// Basic Type Checking Functions
// ------------------------------------------------------------------------------
/**
 * Check if a node is of a specific type.
 *
 * @param node The AST node
 * @param type The type to check for
 * @returns True if the node is of the specified type
 */
function isType(node, type) {
    return node?.type === type;
}
/**
 * Check if a node is an Identifier.
 *
 * @param node The AST node
 * @returns True if the node is an Identifier
 */
function isIdentifier(node) {
    return isType(node, 'Identifier');
}
/**
 * Check if a node is a MemberExpression.
 *
 * @param node The AST node
 * @returns True if the node is a MemberExpression
 */
function isMember(node) {
    return isType(node, 'MemberExpression');
}
/**
 * Check if a node is a CallExpression.
 *
 * @param node The AST node
 * @returns True if the node is a CallExpression
 */
function isCall(node) {
    return isType(node, 'CallExpression');
}
/**
 * Check if a node is a Literal.
 *
 * @param node The AST node
 * @returns True if the node is a Literal
 */
function isLiteral(node) {
    return isType(node, 'Literal');
}
/**
 * Check if a node is a Property.
 *
 * @param node The AST node
 * @returns True if the node is a Property
 */
function isProperty(node) {
    return isType(node, 'Property');
}
/**
 * Check if a node is an ObjectExpression.
 *
 * @param node The AST node
 * @returns True if the node is an ObjectExpression
 */
function isObject(node) {
    return isType(node, 'ObjectExpression');
}
/**
 * Check if a node is an ArrayExpression.
 *
 * @param node The AST node
 * @returns True if the node is an ArrayExpression
 */
function isArray(node) {
    return isType(node, 'ArrayExpression');
}
// ------------------------------------------------------------------------------
// Array Utility Functions
// ------------------------------------------------------------------------------
/**
 * Check if an array contains an item.
 *
 * @param array The array to search in
 * @param item The item to search for
 * @returns True if the array contains the item
 */
function contains(array, item) {
    return array.includes(item);
}
// ------------------------------------------------------------------------------
// String Utility Functions
// ------------------------------------------------------------------------------
/**
 * Check if a value is a string.
 *
 * @param value The value to check
 * @returns True if the value is a string
 */
function isString(value) {
    return typeof value === 'string';
}
/**
 * Check if a string starts with a substring.
 *
 * @param base The base string to check
 * @param sub The substring to look for at the start
 * @returns True if the base string starts with the substring
 */
function startsWith(base, sub) {
    return base.startsWith(sub);
}
/**
 * Check if a string contains a substring.
 *
 * @param str The string to search in
 * @param substring The substring to search for
 * @returns True if the string contains the substring
 */
function containsString(str, substring) {
    return str.includes(substring);
}
// ------------------------------------------------------------------------------
// Window Object Detection Functions
// ------------------------------------------------------------------------------
/**
 * Check if a node represents the global 'window' variable.
 *
 * @param node The AST node
 * @returns True if the node represents the global window variable
 */
function isWindow(node) {
    return !!node && isIdentifier(node) && node.name === 'window';
}
/**
 * Create a function to check if a node is the window object or a reference to it.
 *
 * @param windowObjects Array of window object names to check against
 * @returns Function that checks if a node is a window object or reference
 */
function createIsWindowObject(windowObjects) {
    return function isWindowObject(node) {
        return (isWindow(node) || (!!node && isIdentifier(node) && contains(windowObjects, node.name)));
    };
}
/**
 * Create a function to remember window references.
 *
 * @param windowObjects Array to store window object names
 * @param isWindowObject Function to check if a node is a window object
 * @returns Function that remembers window object references
 */
function createRememberWindow(windowObjects, isWindowObject) {
    return function rememberWindow(left, right) {
        if (isWindowObject(right) && isIdentifier(left)) {
            windowObjects.push(left.name);
            return true;
        }
        return false;
    };
}
// ------------------------------------------------------------------------------
// Document Object Detection Functions
// ------------------------------------------------------------------------------
/**
 * Create a function to check if a node represents the global 'document' variable.
 *
 * @param isWindowObject Function to check if a node is a window object
 * @returns Function that checks if a node represents the document object
 */
function createIsDocument(isWindowObject) {
    return function isDocument(node) {
        if (node) {
            if (isIdentifier(node)) {
                return node.name === 'document';
            }
            else if (isMember(node)) {
                const memberNode = node;
                return isWindowObject(memberNode.object) && isDocument(memberNode.property);
            }
        }
        return false;
    };
}
/**
 * Create a function to check if a node is the document object or a reference to it.
 *
 * @param documentObjects Array to store document object names
 * @param isDocument Function to check if a node is a document object
 * @returns Function that checks if a node is a document object or reference
 */
function createIsDocumentObject(documentObjects, isDocument) {
    return function isDocumentObject(node) {
        return (isDocument(node) ||
            (!!node && isIdentifier(node) && contains(documentObjects, node.name)));
    };
}
/**
 * Create a function to remember document references.
 *
 * @param documentObjects Array to store document object names
 * @param isDocumentObject Function to check if a node is a document object
 * @returns Function that remembers document object references
 */
function createRememberDocument(documentObjects, isDocumentObject) {
    return function rememberDocument(left, right) {
        if (isDocumentObject(right) && isIdentifier(left)) {
            documentObjects.push(left.name);
            return true;
        }
        return false;
    };
}
// ------------------------------------------------------------------------------
// AST Path Building Functions
// ------------------------------------------------------------------------------
/**
 * Build a path string from a callee node.
 *
 * @param node The AST node
 * @returns String representation of the callee path
 */
function buildCalleePath(node) {
    const memberNode = node;
    if (isMember(memberNode.object)) {
        const objectNode = memberNode.object;
        const propertyName = isIdentifier(objectNode.property) ? objectNode.property.name : '';
        return `${buildCalleePath(memberNode.object)}.${propertyName}`;
    }
    else if (isIdentifier(memberNode.object)) {
        return memberNode.object.name;
    }
    return '';
}
/**
 * Convert a MemberExpression to a string representation.
 *
 * @param node The AST node
 * @returns String representation of the member expression
 */
function getMemberAsString(node) {
    if (isMember(node)) {
        const memberNode = node;
        return `${getMemberAsString(memberNode.object)}.${getMemberAsString(memberNode.property)}`;
    }
    else if (isLiteral(node)) {
        return String(node.value);
    }
    else if (isIdentifier(node)) {
        return node.name;
    }
    return '';
}
// ------------------------------------------------------------------------------
// History Object Detection Functions
// ------------------------------------------------------------------------------
/**
 * Create a function to check if a node represents the 'history' object.
 *
 * @param isWindowObject Function to check if a node is a window object
 * @returns Function that checks if a node represents the history object
 */
function createIsHistory(isWindowObject) {
    return function isHistory(node) {
        if (node) {
            if (isIdentifier(node)) {
                return node.name === 'history';
            }
            else if (isMember(node)) {
                const memberNode = node;
                return isWindowObject(memberNode.object) && isHistory(memberNode.property);
            }
        }
        return false;
    };
}
/**
 * Create a function to check if a node is the history object or a reference to it.
 *
 * @param historyObjects Array to store history object names
 * @param isHistory Function to check if a node is a history object
 * @returns Function that checks if a node is a history object or reference
 */
function createIsHistoryObject(historyObjects, isHistory) {
    return function isHistoryObject(node) {
        return (isHistory(node) || (!!node && isIdentifier(node) && contains(historyObjects, node.name)));
    };
}
/**
 * Create a function to remember history references.
 *
 * @param historyObjects Array to store history object names
 * @param isHistoryObject Function to check if a node is a history object
 * @returns Function that remembers history object references
 */
function createRememberHistory(historyObjects, isHistoryObject) {
    return function rememberHistory(left, right) {
        if (isHistoryObject(right) && isIdentifier(left)) {
            historyObjects.push(left.name);
            return true;
        }
        return false;
    };
}
// ------------------------------------------------------------------------------
// Location Object Detection Functions
// ------------------------------------------------------------------------------
/**
 * Create a function to check if a node represents the 'location' object.
 *
 * @param isWindowObject Function to check if a node is a window object
 * @returns Function that checks if a node represents the location object
 */
function createIsLocation(isWindowObject) {
    return function isLocation(node) {
        if (node) {
            if (isIdentifier(node)) {
                return node.name === 'location';
            }
            else if (isMember(node)) {
                const memberNode = node;
                return isWindowObject(memberNode.object) && isLocation(memberNode.property);
            }
        }
        return false;
    };
}
/**
 * Create a function to check if a node is the location object or a reference to it.
 *
 * @param locationObjects Array to store location object names
 * @param isLocation Function to check if a node is a location object
 * @returns Function that checks if a node is a location object or reference
 */
function createIsLocationObject(locationObjects, isLocation) {
    return function isLocationObject(node) {
        return (isLocation(node) ||
            (!!node && isIdentifier(node) && contains(locationObjects, node.name)));
    };
}
/**
 * Create a function to remember location references.
 *
 * @param locationObjects Array to store location object names
 * @param isLocationObject Function to check if a node is a location object
 * @returns Function that remembers location object references
 */
function createRememberLocation(locationObjects, isLocationObject) {
    return function rememberLocation(left, right) {
        if (isLocationObject(right) && isIdentifier(left)) {
            locationObjects.push(left.name);
            return true;
        }
        return false;
    };
}
// ------------------------------------------------------------------------------
// Common Utility Functions for Rule Processing
// ------------------------------------------------------------------------------
/**
 * Get the literal or identifier name from a node.
 *
 * @param node The AST node
 * @returns The name or value as a string
 */
function getLiteralOrIdentifierName(node) {
    if (isIdentifier(node)) {
        return node.name;
    }
    else if (isLiteral(node)) {
        return String(node.value);
    }
    return '';
}
/**
 * Get the identifier path from a node, handling various node types.
 *
 * @param node The AST node
 * @returns The identifier path as a string
 */
function getIdentifierPath(node) {
    if (!node) {
        return '';
    }
    const astNode = node;
    switch (astNode.type) {
        case 'Identifier':
            return node.name;
        case 'MemberExpression': {
            const memberNode = node;
            return `${getIdentifierPath(memberNode.object)}.${getLiteralOrIdentifierName(memberNode.property)}`;
        }
        case 'NewExpression':
            return getIdentifierPath(node.callee);
        case 'CallExpression':
            return `${getIdentifierPath(node.callee)}().`;
        default:
            return '';
    }
}
/**
 * Resolve identifier path by substituting known variables.
 *
 * @param path The path to resolve
 * @param variables The variables lookup object
 * @returns The resolved path with substitutions applied
 */
function resolveIdentifierPath(path, variables) {
    const parts = path.split('.');
    let substitution;
    // Check if current identifier is remembered as an interesting variable
    for (const name in variables) {
        if (name === parts[0]) {
            // Get last stored variable value
            substitution = variables[name].at(-1);
        }
    }
    // If so, replace current identifier with its value
    if (substitution) {
        parts[0] = substitution;
        path = parts.join('.');
    }
    return path;
}
/**
 * Remember interesting variable for later resolution.
 *
 * @param node The variable declarator node
 * @param name The name to remember
 * @param variables The variables lookup object
 */
function rememberInterestingVariable(node, name, variables) {
    const declaratorNode = node;
    variables[declaratorNode.id.name] ??= [];
    variables[declaratorNode.id.name].push(name);
}
/**
 * Check if a path is interesting based on a given criteria function.
 *
 * @param path The path to check
 * @param interestingPathChecker Function to determine if path is interesting
 * @returns True if the path is considered interesting
 */
function checkInterestingPath(path, interestingPathChecker) {
    return interestingPathChecker(path);
}
/**
 * Create a generic variable declarator processor.
 *
 * @param variables The variables lookup object
 * @param interestingPathChecker Function to determine if path is interesting
 * @returns Function that processes variable declarator nodes
 */
function createVariableDeclaratorProcessor(variables, interestingPathChecker) {
    return function processVariableDeclarator(node) {
        const declaratorNode = node;
        let path = getIdentifierPath(declaratorNode.init);
        path = resolveIdentifierPath(path, variables);
        // If declaration is interesting, remember identifier and resolved value
        if (checkInterestingPath(path, interestingPathChecker)) {
            rememberInterestingVariable(node, path, variables);
        }
    };
}
/**
 * Check if an identifier has an underscore prefix (indicating private).
 *
 * @param identifier The identifier to check
 * @returns True if the identifier has an underscore prefix
 */
function hasUnderscore(identifier) {
    return identifier !== '_' && identifier.startsWith('_');
}
/**
 * Check if a value is an integer.
 *
 * @param value The value to check
 * @returns True if the value is an integer
 */
function isInteger(value) {
    return Number(value) === value && value % 1 === 0;
}
/**
 * Check if a string ends with a specific substring.
 *
 * @param str The string to check
 * @param suffix The suffix to look for
 * @returns True if the string ends with the suffix
 */
function endsWith(str, suffix) {
    return typeof str === 'string' && typeof suffix === 'string' && str.endsWith(suffix);
}
/**
 * Get the last element of a dot-separated path.
 *
 * @param calleePath The path to extract from
 * @returns The last element of the path
 */
function isForbiddenObviousApi(calleePath) {
    const elementArray = calleePath.split('.');
    return elementArray.at(-1) ?? '';
}
// ------------------------------------------------------------------------------
// Storage Detection Functions (for localStorage/sessionStorage rules)
// ------------------------------------------------------------------------------
/**
 * Create a storage rule factory that generates consistent rules for localStorage/sessionStorage.
 *
 * @param storageName The name of the storage API (e.g., 'localStorage' or 'sessionStorage')
 * @param messageId The message ID for the rule violation
 * @returns Object containing helper functions for storage detection
 */
function createStorageRuleHelpers(storageName, messageId) {
    return {
        /**
         * Process variable declarator nodes for storage references.
         *
         * @param node The variable declarator node to process
         * @param forbiddenStorageObjects Array to store forbidden storage object references
         */
        processVariableDeclarator(node, forbiddenStorageObjects) {
            const declaratorNode = node;
            if (declaratorNode.init) {
                if (declaratorNode.init.type === 'MemberExpression') {
                    const memberInit = declaratorNode.init;
                    const objectNode = memberInit.object;
                    const propertyNode = memberInit.property;
                    const firstElement = objectNode.name;
                    const secondElement = propertyNode.name;
                    if (firstElement + '.' + secondElement === `window.${storageName}`) {
                        forbiddenStorageObjects.push(declaratorNode.id.name);
                    }
                }
                else if (declaratorNode.init.type === 'Identifier' &&
                    declaratorNode.init.name === storageName) {
                    forbiddenStorageObjects.push(declaratorNode.id.name);
                }
            }
        },
        /**
         * Check if a member expression represents forbidden storage usage.
         *
         * @param node The member expression node to check
         * @param forbiddenStorageObjects Array of forbidden storage object references
         * @param context The ESLint rule context
         */
        checkMemberExpression(node, forbiddenStorageObjects, context) {
            const memberExpressionNode = node;
            const calleePath = buildCalleePath(memberExpressionNode);
            const speciousObject = isForbiddenObviousApi(calleePath);
            if (((calleePath === storageName || calleePath === `window.${storageName}`) &&
                speciousObject === storageName) ||
                contains(forbiddenStorageObjects, speciousObject)) {
                context.report({ node: node, messageId });
            }
        }
    };
}
// ------------------------------------------------------------------------------
// Property Checking Functions (for UI5 property rules)
// ------------------------------------------------------------------------------
/**
 * Create a helper function to check member expressions against forbidden property lists.
 *
 * @param propertyGroups Object mapping message IDs to arrays of forbidden property names
 * @returns Function that checks member expressions
 */
function createPropertyChecker(propertyGroups) {
    return function checkMemberExpression(node, context) {
        const memberNode = node;
        if (!memberNode.property) {
            return;
        }
        const propertyObj = memberNode.property;
        if (!('name' in propertyObj)) {
            return;
        }
        const val = memberNode.property.name;
        if (typeof val === 'string') {
            for (const [messageId, properties] of Object.entries(propertyGroups)) {
                if (contains(properties, val)) {
                    const data = messageId.includes('Prop') ? { property: val } : undefined;
                    context.report({ node, messageId, data });
                    break;
                }
            }
        }
    };
}
// ------------------------------------------------------------------------------
// Document-based Rule Functions (for DOM access, element creation, etc.)
// ------------------------------------------------------------------------------
/**
 * Create document-based rule visitors with common pattern.
 *
 * @param config Configuration object for the document-based rule
 * @param config.isInteresting Function to check if a node is interesting
 * @param config.isValid Function to validate the node
 * @param config.messageId The message ID for violations
 * @returns Rule visitor object
 */
function createDocumentBasedRuleVisitors(config) {
    return function createVisitors(context) {
        const WINDOW_OBJECTS = [];
        const DOCUMENT_OBJECTS = [];
        // Initialize factory functions
        const isWindowObject = createIsWindowObject(WINDOW_OBJECTS);
        const rememberWindow = createRememberWindow(WINDOW_OBJECTS, isWindowObject);
        const isDocument = createIsDocument(isWindowObject);
        const isDocumentObject = createIsDocumentObject(DOCUMENT_OBJECTS, isDocument);
        const rememberDocument = createRememberDocument(DOCUMENT_OBJECTS, isDocumentObject);
        return {
            'VariableDeclarator'(node) {
                return (rememberWindow(node.id, node.init) ||
                    rememberDocument(node.id, node.init));
            },
            'AssignmentExpression'(node) {
                return (rememberWindow(node.left, node.right) ||
                    rememberDocument(node.left, node.right));
            },
            'MemberExpression'(node) {
                if (config.isInteresting(node, isDocumentObject) && !config.isValid(node)) {
                    context.report({ node: node, messageId: config.messageId });
                }
            }
        };
    };
}
/**
 * Validates that a path exists and returns the deepest existing level.
 *
 * @param startObject - The object to start validation from (e.g., settings object)
 * @param pathSegments - Array of path segments
 * @returns Object with validatedPath (deepest existing) and missingSegments
 */
function findDeepestExistingPath(startObject, pathSegments) {
    let current = startObject;
    for (let i = 0; i < pathSegments.length; i++) {
        const segment = pathSegments[i];
        if (!current || typeof current !== 'object' || !(segment in current)) {
            return {
                validatedPath: pathSegments.slice(0, i),
                missingSegments: pathSegments.slice(i)
            };
        }
        current = current[segment];
    }
    return {
        validatedPath: pathSegments,
        missingSegments: []
    };
}
