// @sap-ux/eslint-plugin-fiori-tools/lib/project-context/parser/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationParser = void 0;
const node_path_1 = require("node:path");
const node_url_1 = require("node:url");
const momoa_1 = require("@humanwhocodes/momoa");
const project_access_1 = require("@sap-ux/project-access");
const fiori_annotation_api_1 = require("@sap-ux/fiori-annotation-api");
const utils_1 = require("../utils");
const service_1 = require("./service");
/**
 *
 */
class ApplicationParser {
    context;
    index;
    diagnostics;
    /**
     * Resets the parser state with new project type and file cache.
     *
     * @param projectType - The type of project being parsed
     * @param fileCache - Map of file URIs to their contents
     */
    reset(projectType, fileCache) {
        this.index = { projectType, apps: {}, documents: {} };
        this.diagnostics = [];
        this.context = { projectType, fileCache };
    }
    /**
     *
     * @param projectType
     * @param artifacts
     * @param fileCache
     */
    parse(projectType, artifacts, fileCache) {
        this.reset(projectType, fileCache);
        for (const app of artifacts.applications ?? []) {
            try {
                const manifestUri = (0, node_url_1.pathToFileURL)(app.manifestPath).toString();
                const manifestContent = fileCache.get(manifestUri) ?? '';
                const manifestAst = (0, momoa_1.parse)(manifestContent, {
                    mode: 'json',
                    ranges: true,
                    tokens: true,
                    allowTrailingCommas: false
                });
                this.index.documents[manifestUri] = manifestAst;
                const manifest = JSON.parse(manifestContent);
                const webappPath = (0, node_path_1.dirname)(app.manifestPath);
                const [parsedManifest, services] = this.parseManifest(webappPath, manifestUri, manifest);
                const appRootUri = (0, node_url_1.pathToFileURL)(app.appRoot).toString();
                const parsedApp = {
                    manifest: parsedManifest,
                    manifestObject: manifest,
                    projectRootPath: app.projectRoot,
                    services: {}
                };
                this.index.apps[appRootUri] = parsedApp;
                for (const service of services) {
                    const artifacts = this.parseService(app.projectRoot, service);
                    if (artifacts) {
                        const index = (0, service_1.buildServiceIndex)(artifacts, this.index.documents);
                        parsedApp.services[service.name] = { config: service, artifacts, index };
                    }
                }
            }
            catch {
                // skip faulty apps for now
            }
        }
        return {
            index: this.index,
            diagnostics: this.diagnostics
        };
    }
    /**
     * Reparses CDS files and updates the project index.
     *
     * @param index - The current parsed project index
     * @param fileCache - Map of file URIs to their contents
     */
    reparseCDS(index, fileCache) {
        // there is assumption that only one eslint config exists in CAP project and its run for the whole project
        const projectRoot = index.apps[Object.keys(index.apps)[0]]?.projectRootPath;
        if (projectRoot) {
            fiori_annotation_api_1.CdsAnnotationProvider.resetCache(projectRoot, fileCache);
        }
        for (const app of Object.values(index.apps)) {
            for (const service of Object.values(app.services)) {
                if (service.config.type === 'cap') {
                    const artifacts = this.parseService(app.projectRootPath, service.config);
                    if (artifacts) {
                        const serviceIndex = (0, service_1.buildServiceIndex)(artifacts, index.documents);
                        service.artifacts = artifacts;
                        service.index = serviceIndex;
                    }
                }
            }
        }
    }
    /**
     * Reparses a manifest.json file and updates the associated app in the project index.
     *
     * @param uri - The URI of the manifest.json file to reparse
     * @param index - The current parsed project index
     * @param fileCache - Map of file URIs to their contents
     */
    reparseJSON(uri, index, fileCache) {
        for (const [key, previousApp] of Object.entries(index.apps)) {
            if (previousApp.manifest.manifestUri !== uri) {
                continue;
            }
            const manifestContent = fileCache.get(uri) ?? '';
            const manifestAst = (0, momoa_1.parse)(manifestContent, {
                mode: 'json',
                ranges: true,
                tokens: true,
                allowTrailingCommas: false
            });
            index.documents[uri] = manifestAst;
            const manifest = JSON.parse(manifestContent);
            const webappPath = (0, node_path_1.dirname)((0, node_url_1.fileURLToPath)(uri));
            const [parsedManifest, services] = this.parseManifest(webappPath, uri, manifest);
            index.documents[uri] = manifestAst;
            const parsedApp = {
                manifest: parsedManifest,
                manifestObject: manifest,
                projectRootPath: previousApp.projectRootPath,
                services: {}
            };
            const previouslyFoundServices = Object.values(previousApp.services).map((service) => service.config);
            if (JSON.stringify(previouslyFoundServices) === JSON.stringify(services)) {
                // keep existing services
                parsedApp.services = previousApp.services;
            }
            else {
                // services have changed, reset existing services
                parsedApp.services = {};
                for (const service of services) {
                    const artifacts = this.parseService(parsedApp.projectRootPath, service);
                    if (artifacts) {
                        const serviceIndex = (0, service_1.buildServiceIndex)(artifacts, index.documents);
                        parsedApp.services[service.name] = { config: service, artifacts, index: serviceIndex };
                    }
                }
            }
            index.apps[key] = parsedApp;
            break;
        }
    }
    /**
     * Reparses an XML annotation file and updates the affected services in the project index.
     *
     * @param uri - The URI of the XML file to reparse
     * @param index - The current parsed project index
     */
    reparseXML(uri, index) {
        for (const app of Object.values(index.apps)) {
            for (const service of Object.values(app.services)) {
                if (service.config.type === 'local') {
                    const annotationFile = service.config.annotationFiles.find((file) => file.uri === uri);
                    if (annotationFile) {
                        const artifacts = this.parseService(app.projectRootPath, service.config);
                        if (artifacts) {
                            const serviceIndex = (0, service_1.buildServiceIndex)(artifacts, index.documents);
                            service.artifacts = artifacts;
                            service.index = serviceIndex;
                        }
                        break;
                    }
                }
            }
        }
    }
    /**
     * Reparses a specific file and updates the project index based on file type.
     *
     * @param uri - The URI of the file to reparse
     * @param index - The current parsed project index
     * @param fileCache - Map of file URIs to their contents
     */
    reparse(uri, index, fileCache) {
        this.reset(index.projectType, fileCache);
        if (uri.endsWith('.cds')) {
            this.reparseCDS(index, fileCache);
        }
        else if (uri.endsWith('manifest.json')) {
            this.reparseJSON(uri, index, fileCache);
        }
        else if (uri.endsWith('.xml')) {
            this.reparseXML(uri, index);
        }
        return { index: index, diagnostics: [] };
    }
    /**
     * Parses a manifest.json file to extract app configuration and OData services.
     *
     * @param webappPath - The absolute path to the webapp directory
     * @param manifestUri - The URI of the manifest.json file
     * @param manifest - The parsed manifest object
     */
    parseManifest(webappPath, manifestUri, manifest) {
        const customViews = {};
        const services = [];
        const targets = manifest['sap.ui5']?.routing?.targets;
        for (const [, target] of Object.entries(targets ?? {})) {
            const settings = target.options?.settings;
            if (settings?.entitySet || settings?.contextPath) {
                if (settings.viewName) {
                    customViews[settings.viewName] = {
                        entitySet: settings.entitySet,
                        contextPath: settings.contextPath
                    };
                }
            }
        }
        const mainServiceName = (0, project_access_1.getMainService)(manifest) ?? '';
        const manifestDataSources = manifest['sap.app']?.dataSources ?? {};
        for (const [dataSourceName, dataSource] of Object.entries(manifestDataSources)) {
            if (dataSource.type !== 'OData' || !dataSource.uri) {
                continue;
            }
            const localUri = dataSource.settings?.localUri;
            if (!localUri) {
                // no local metadata, assume cap service
                services.push({
                    type: 'cap',
                    name: dataSourceName,
                    path: (0, utils_1.uniformUrl)(dataSource.uri),
                    version: dataSource.settings?.odataVersion ?? '2.0'
                });
                continue;
            }
            const metadataPath = (0, node_path_1.join)(webappPath, localUri);
            const metadataUri = (0, node_url_1.pathToFileURL)(metadataPath).toString();
            const annotationFiles = getAnnotationFiles(webappPath, dataSource, manifestDataSources);
            services.push({
                type: 'local',
                name: dataSourceName,
                path: (0, utils_1.uniformUrl)(dataSource.uri),
                version: dataSource.settings?.odataVersion ?? '2.0',
                metadata: {
                    type: 'remote',
                    cacheType: 'local-service',
                    cachePath: metadataPath,
                    relativeBackendPath: `${dataSource.uri}$metadata`,
                    uri: metadataUri
                },
                annotationFiles
            });
        }
        const flexEnabled = getFlexEnabled(manifest);
        const minUI5Version = getMinUI5Version(manifest);
        const parsedManifest = {
            manifestUri,
            webappPath,
            flexEnabled,
            minUI5Version,
            customViews,
            appId: manifest['sap.app']?.id ?? '',
            mainServiceName
        };
        return [parsedManifest, services];
    }
    /**
     * Parses an OData service and returns its artifacts (metadata and annotations).
     *
     * @param projectRootPath - The absolute path to the project root
     * @param service - The OData service configuration to parse
     */
    parseService(projectRootPath, service) {
        if (service.type === 'cap') {
            return fiori_annotation_api_1.CdsAnnotationProvider.getCdsServiceArtifacts(projectRootPath, service.path, this.context.fileCache);
        }
        else {
            return (0, fiori_annotation_api_1.getXmlServiceArtifacts)(service.version, service.path, { uri: service.metadata.uri, isReadOnly: true }, service.annotationFiles.map((file) => ({ uri: file.uri, isReadOnly: file.type === 'remote' })), this.context.fileCache);
        }
    }
    /**
     * Checks if the current project is a CAP (Cloud Application Programming) project.
     */
    isCapProject() {
        return this.context.projectType === 'CAPJava' || this.context.projectType === 'CAPNodejs';
    }
}
exports.ApplicationParser = ApplicationParser;
/**
 * Retrieves the list of annotation files configured for an OData data source.
 *
 * @param webappPath - The absolute path to the webapp directory
 * @param dataSource - The OData data source configuration
 * @param manifestDataSources - All data sources defined in the manifest
 */
function getAnnotationFiles(webappPath, dataSource, manifestDataSources) {
    const annotationFiles = [];
    for (const id of dataSource.settings?.annotations ?? []) {
        const annotationDataSource = manifestDataSources[id];
        if (annotationDataSource.type !== 'ODataAnnotation') {
            continue;
        }
        const { settings, uri } = annotationDataSource;
        const localUri = settings?.localUri ?? '';
        const filePath = (0, node_path_1.join)(webappPath, localUri);
        const annotationFileUri = (0, node_url_1.pathToFileURL)(filePath).toString();
        const isReadonly = !(uri && localUri && localUri === uri);
        if (isReadonly) {
            annotationFiles.push({
                type: 'remote',
                cacheType: 'local-service',
                cachePath: filePath,
                relativeBackendPath: uri,
                uri: annotationFileUri
            });
        }
        else {
            annotationFiles.push({
                type: 'local',
                uri: annotationFileUri
            });
        }
    }
    return annotationFiles;
}
/**
 * Determines if SAPUI5 flexibility (UI adaptation) is enabled in the manifest.
 *
 * @param manifest - The parsed manifest object
 */
function getFlexEnabled(manifest) {
    return manifest['sap.ui5']?.flexEnabled ?? false;
}
/**
 * Extracts and parses the minimum UI5 version required by the application.
 *
 * @param manifest - The parsed manifest object
 */
function getMinUI5Version(manifest) {
    const value = manifest['sap.ui5']?.dependencies?.minUI5Version;
    if (!value) {
        return undefined;
    }
    const rawValue = Array.isArray(value) ? value[0] : value;
    if (rawValue.startsWith('${') && rawValue.endsWith('}')) {
        // expression, assume latest UI5 version
        return {
            raw: rawValue,
            major: 999,
            minor: 0,
            patch: 0
        };
    }
    const [major, minor, patch] = rawValue.split('.').map((part) => Number.parseInt(part, 10));
    return {
        raw: rawValue,
        major: Number.isNaN(major) ? 0 : major,
        minor: Number.isNaN(minor) ? 0 : minor,
        patch: Number.isNaN(patch) ? 0 : patch
    };
}
