// @sap-ux/eslint-plugin-fiori-tools/lib/project-context/parser/service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildAnnotationIndexKey = buildAnnotationIndexKey;
exports.buildServiceIndex = buildServiceIndex;
const odata_annotation_core_1 = require("@sap-ux/odata-annotation-core");
/**
 * Builds an annotation index key from target and term.
 *
 * @param target - The annotation target
 * @param term - The annotation term
 * @returns The index key in format "target/@term"
 */
function buildAnnotationIndexKey(target, term) {
    return `${target}/@${term}`;
}
/**
 * Indexes annotations by their annotation path (target and term).
 *
 * @param service - Service artifacts containing annotation files
 * @returns Annotation index organized by path and qualifier
 */
function indexAnnotationsByAnnotationPath(service) {
    const index = {};
    for (const file of service.fileSequence) {
        const annotationFile = service.annotationFiles[file];
        if (!annotationFile) {
            continue;
        }
        const namespace = annotationFile.namespace ?? { name: 'DEFAULT_NAMESPACE', type: 'namespace' };
        const namespaces = (0, odata_annotation_core_1.getAllNamespacesAndReferences)(namespace, annotationFile.references);
        const aliasInfo = (0, odata_annotation_core_1.getAliasInformation)(namespaces, service.metadataService.getNamespaces());
        for (const target of annotationFile.targets) {
            const targetName = (0, odata_annotation_core_1.toFullyQualifiedName)(aliasInfo.aliasMap, namespace.name, (0, odata_annotation_core_1.parseIdentifier)(target.name));
            if (!targetName) {
                continue;
            }
            processTargetAnnotations(target, annotationFile, namespace, aliasInfo, targetName, index);
        }
    }
    return index;
}
/**
 * Processes annotations for a specific target and adds them to the index.
 *
 * @param target - The annotation target to process
 * @param annotationFile - The annotation file containing the target
 * @param namespace - Namespace information
 * @param aliasInfo - Alias information for resolving qualified names
 * @param targetName - Fully qualified target name
 * @param index - Annotation index to update
 */
function processTargetAnnotations(target, annotationFile, namespace, aliasInfo, targetName, index) {
    for (const annotation of target.terms) {
        const term = (0, odata_annotation_core_1.getElementAttributeValue)(annotation, "Term" /* Edm.Term */);
        const qualifier = (0, odata_annotation_core_1.getElementAttribute)(annotation, "Qualifier" /* Edm.Qualifier */);
        const termName = (0, odata_annotation_core_1.toFullyQualifiedName)(aliasInfo.aliasMap, namespace.name, (0, odata_annotation_core_1.parseIdentifier)(term));
        if (!termName) {
            continue;
        }
        const termKey = buildAnnotationIndexKey(targetName, termName);
        index[termKey] ??= {};
        const qualifierKey = qualifier?.value ?? 'undefined';
        const indexedValue = index[termKey][qualifierKey];
        const reference = {
            uri: annotationFile.uri,
            value: annotation
        };
        if (indexedValue) {
            indexedValue.top = reference;
            indexedValue.layers.unshift(reference);
        }
        else {
            index[termKey][qualifierKey] = {
                source: annotationFile.uri,
                target: targetName,
                term: termName,
                qualifier: qualifier?.value,
                top: reference,
                layers: [reference]
            };
        }
    }
}
/**
 * Builds a service index from service artifacts.
 * Creates indexes for entity sets, entity containers, and annotations.
 *
 * @param artifacts - Service artifacts to index
 * @param documents - Document map to populate with annotation files
 * @returns Complete service index with entity sets and annotations
 */
function buildServiceIndex(artifacts, documents) {
    const entitySets = {};
    for (const document of Object.values(artifacts.annotationFiles)) {
        documents[document.uri] = document;
    }
    const annotationIndex = indexAnnotationsByAnnotationPath(artifacts);
    let entityContainer;
    artifacts.metadataService.visitMetadataElements((element) => {
        // NOSONAR - TODO: check if we can handle CDS differences better
        if (element.kind === 'EntityContainer' || element.kind === 'service') {
            entityContainer = element;
        }
        else if (element.kind === 'EntitySet' || element.kind === 'entitySet') {
            entitySets[element.name] = element;
        }
    });
    return {
        entitySets: entitySets,
        entityContainer,
        annotations: annotationIndex
    };
}
