// @sap-ux/eslint-plugin-fiori-tools/lib/project-context/linker/annotations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigurationKey = getConfigurationKey;
exports.collectTables = collectTables;
exports.collectSections = collectSections;
exports.getRecordType = getRecordType;
exports.getEntityForContextPath = getEntityForContextPath;
exports.collectHeaderSections = collectHeaderSections;
const odata_annotation_core_1 = require("@sap-ux/odata-annotation-core");
const parser_1 = require("../parser");
const constants_1 = require("../../constants");
/**
 * Creates a configuration key from an annotation path
 *
 * @param annotationPath
 */
function getConfigurationKey(annotationPath) {
    return annotationPath
        .split('/')
        .map((segment) => segment.replace('@', ''))
        .join('::');
}
/**
 * Collects table nodes from UI.LineItem annotations for an entity type.
 *
 * @param feVersion - The Fiori Elements version ('v2' or 'v4')
 * @param entityType - The entity type name
 * @param service - The parsed OData service
 */
function collectTables(feVersion, entityType, service) {
    const tables = [];
    const lineItemKey = (0, parser_1.buildAnnotationIndexKey)(entityType, constants_1.UI_LINE_ITEM);
    const lineItems = service.index.annotations[lineItemKey]?.['undefined'];
    if (lineItems) {
        const table = {
            type: 'table',
            annotation: lineItems,
            annotationPath: `@com.sap.vocabularies.UI.v1.LineItem`,
            children: []
        };
        tables.push(table);
    }
    return tables;
}
/**
 * Collects section nodes from UI.Facets an UI.HeaderFacet annotations for an entity type.
 *
 * @param feVersion - The Fiori Elements version ('v2' or 'v4')
 * @param entityType - The entity type name
 * @param service - The parsed OData service
 */
function collectSections(feVersion, entityType, service) {
    const sections = [];
    const facetsKey = (0, parser_1.buildAnnotationIndexKey)(entityType, 'com.sap.vocabularies.UI.v1.Facets');
    const facets = service.index.annotations[facetsKey]?.['undefined'];
    if (!facets) {
        return sections;
    }
    const [collection] = (0, odata_annotation_core_1.elementsWithName)("Collection" /* Edm.Collection */, facets.top.value);
    if (!collection) {
        return sections;
    }
    const records = (0, odata_annotation_core_1.elementsWithName)("Record" /* Edm.Record */, collection);
    const aliasInfo = service.artifacts.aliasInfo[facets.top.uri];
    let index = 0;
    for (const record of records) {
        const section = processReferenceFacetRecord(record, aliasInfo, entityType, service, facets, index);
        if (section) {
            sections.push(section);
        }
        index++;
    }
    const headerFacetsKey = (0, parser_1.buildAnnotationIndexKey)(entityType, 'com.sap.vocabularies.UI.v1.HeaderFacets');
    const headerFacets = service.index.annotations[headerFacetsKey]?.['undefined'];
    if (!headerFacets) {
        return sections;
    }
    const [headerFacetCollection] = (0, odata_annotation_core_1.elementsWithName)("Collection" /* Edm.Collection */, headerFacets.top.value);
    if (!headerFacetCollection) {
        return sections;
    }
    const headerFacetRecords = (0, odata_annotation_core_1.elementsWithName)("Record" /* Edm.Record */, headerFacetCollection);
    const headerFacetAliasInfo = service.artifacts.aliasInfo[headerFacets.top.uri];
    index = 0;
    for (const record of headerFacetRecords) {
        const headerFacet = processReferenceFacetRecord(record, headerFacetAliasInfo, entityType, service, headerFacets, index);
        if (headerFacet) {
            sections.push(headerFacet);
        }
        index++;
    }
    return sections;
}
/**
 * Process a single reference facet record and create a table or header section if applicable.
 *
 * @param record
 * @param aliasInfo
 * @param entityType
 * @param service
 * @param facets
 * @param index
 */
function processReferenceFacetRecord(record, aliasInfo, entityType, service, facets, index) {
    const type = getRecordType(aliasInfo, record);
    if (type !== 'com.sap.vocabularies.UI.v1.ReferenceFacet') {
        return undefined;
    }
    const properties = getRecordPropertyValue(record);
    const id = properties['ID']?.value;
    const target = properties['Target'];
    if (!id || target?.kind !== "AnnotationPath" /* Edm.AnnotationPath */) {
        return undefined;
    }
    const annotationPath = target.value;
    if (annotationPath.startsWith('/')) {
        // absolute path is not supported
        return undefined;
    }
    const referencedEntityType = getReferencedEntityType(aliasInfo, entityType, annotationPath, service);
    if (!referencedEntityType) {
        return undefined;
    }
    const fullyQualifiedPath = (0, odata_annotation_core_1.toFullyQualifiedPath)(aliasInfo.aliasMap, '', (0, odata_annotation_core_1.parsePath)(`/${entityType}/${annotationPath}`));
    const [, _annotationPath] = fullyQualifiedPath.split('@');
    const [term, qualifier] = _annotationPath.split('#');
    if (term === constants_1.UI_LINE_ITEM) {
        return createTableSection(facets, index, referencedEntityType, qualifier, annotationPath, aliasInfo, service);
    }
    if (term === constants_1.UI_FIELD_GROUP) {
        return addHeaderSection(facets, index, referencedEntityType, qualifier, annotationPath, aliasInfo, service);
    }
    return undefined;
}
/**
 * Get the referenced entity type from an annotation path.
 *
 * @param aliasInfo
 * @param entityType
 * @param annotationPath
 * @param service
 */
function getReferencedEntityType(aliasInfo, entityType, annotationPath, service) {
    const fullyQualifiedPath = (0, odata_annotation_core_1.toFullyQualifiedPath)(aliasInfo.aliasMap, '', (0, odata_annotation_core_1.parsePath)(`/${entityType}/${annotationPath}`));
    let [contextPath] = fullyQualifiedPath.split('@');
    if (contextPath.endsWith('/')) {
        contextPath = contextPath.slice(0, -1);
    }
    const referencedEntity = service.artifacts.metadataService.getMetadataElement(contextPath.slice(1));
    return referencedEntity?.kind === 'EntityType' ? referencedEntity.name : referencedEntity?.structuredType;
}
/**
 * Create a table section node with its child table node.
 *
 * @param facets
 * @param index
 * @param referencedEntityType
 * @param qualifier
 * @param annotationPath
 * @param aliasInfo
 * @param service
 */
function createTableSection(facets, index, referencedEntityType, qualifier, annotationPath, aliasInfo, service) {
    const section = {
        type: 'table-section',
        annotationPath: `@com.sap.vocabularies.UI.v1.Facets/${index}`,
        annotation: facets,
        children: []
    };
    const lineItemKey = (0, parser_1.buildAnnotationIndexKey)(referencedEntityType, constants_1.UI_LINE_ITEM);
    const tableAnnotations = service.index.annotations[lineItemKey];
    if (!tableAnnotations) {
        return undefined;
    }
    const annotation = tableAnnotations[qualifier ?? 'undefined'];
    if (!annotation) {
        return undefined;
    }
    const table = {
        type: 'table',
        annotationPath: (0, odata_annotation_core_1.toFullyQualifiedPath)(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, (0, odata_annotation_core_1.parsePath)(annotationPath)),
        annotation,
        children: []
    };
    section.children.push(table);
    return section;
}
/**
 * Creates a header facet section node with field group child annotation.
 *
 * @param headerFacets - Header facet annotation
 * @param index - Index of annotation
 * @param referencedEntityType - Entity type
 * @param qualifier - FieldGroup annotation qualifier
 * @param annotationPath - Header facet annotation path
 * @param aliasInfo - Alias information for resolving namespaces
 * @param service - The parsed OData service
 * @returns Header section annotation node
 */
function addHeaderSection(headerFacets, index, referencedEntityType, qualifier, annotationPath, aliasInfo, service) {
    const section = {
        type: 'header-section',
        annotationPath: `@com.sap.vocabularies.UI.v1.HeaderFacet/${index}`,
        annotation: headerFacets,
        children: []
    };
    const fieldGroupKey = (0, parser_1.buildAnnotationIndexKey)(referencedEntityType, constants_1.UI_FIELD_GROUP);
    const fieldGroupAnnotations = service.index.annotations[fieldGroupKey];
    if (!fieldGroupAnnotations) {
        return undefined;
    }
    const annotation = fieldGroupAnnotations[qualifier ?? 'undefined'];
    if (!annotation) {
        return undefined;
    }
    const fieldGroup = {
        type: 'field-group',
        annotationPath: (0, odata_annotation_core_1.toFullyQualifiedPath)(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, (0, odata_annotation_core_1.parsePath)(annotationPath)),
        annotation,
        children: []
    };
    section.children.push(fieldGroup);
    return section;
}
/**
 * Extracts the record type from an element with alias resolution.
 *
 * @param aliasInfo - Alias information for resolving namespaces
 * @param element - The XML element to extract the type from
 */
function getRecordType(aliasInfo, element) {
    const recordType = (0, odata_annotation_core_1.getElementAttributeValue)(element, "Type" /* Edm.Type */);
    if (recordType.includes('/')) {
        // do not support paths as types
        return undefined;
    }
    if (recordType) {
        return (0, odata_annotation_core_1.toFullyQualifiedName)(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, (0, odata_annotation_core_1.parseIdentifier)(recordType));
    }
}
/**
 * Extracts property values from a record element.
 *
 * @param record - The record element to extract properties from
 */
function getRecordPropertyValue(record) {
    const properties = {};
    for (const child of record.content) {
        if (child.type !== odata_annotation_core_1.ELEMENT_TYPE) {
            continue;
        }
        if (child.name === "PropertyValue" /* Edm.PropertyValue */) {
            const name = (0, odata_annotation_core_1.getElementAttributeValue)(child, "Property" /* Edm.Property */);
            const annotationPathAttribute = (0, odata_annotation_core_1.getElementAttribute)(child, "AnnotationPath" /* Edm.AnnotationPath */);
            if (annotationPathAttribute) {
                properties[name] = {
                    name,
                    value: annotationPathAttribute.value,
                    kind: "AnnotationPath" /* Edm.AnnotationPath */
                };
                continue;
            }
            const stringAttribute = (0, odata_annotation_core_1.getElementAttribute)(child, "String" /* Edm.String */);
            if (stringAttribute) {
                properties[name] = {
                    name,
                    value: stringAttribute.value,
                    kind: "String" /* Edm.String */
                };
            }
        }
    }
    return properties;
}
/**
 * Resolves a metadata element from a context path string.
 *
 * @param contextPath - The context path (e.g., '/EntitySet/NavigationProperty')
 * @param service - The parsed OData service
 */
function getEntityForContextPath(contextPath, service) {
    if (!contextPath.startsWith('/')) {
        return;
    }
    const path = contextPath.substring(1);
    const [entityName, ...segments] = path.split('/');
    if (!entityName) {
        return;
    }
    const entity = service.index.entitySets[entityName];
    if (!entity) {
        return undefined;
    }
    return resolveNavigationProperties(entity, segments);
}
/**
 * Resolves navigation properties along a path to find the target entity.
 *
 * @param root - The starting metadata element
 * @param segments - Array of navigation property names to traverse
 */
function resolveNavigationProperties(root, segments) {
    if (segments.length === 0) {
        return root;
    }
    let current = root;
    for (const segment of segments) {
        let found = false;
        for (const child of current.content) {
            if (child.name === segment) {
                current = child;
                found = true;
                break;
            }
        }
        if (!found) {
            return undefined;
        }
    }
    return current;
}
/**
 * Links an object page header section with its field group annotation.
 * Used by both FE v2 and v4 linkers.
 *
 * @param section - Header section node to link
 * @param page - The object page being linked
 */
function collectHeaderSections(section, page) {
    if (section.type !== 'header-section') {
        return;
    }
    const fieldGroup = section.children[0];
    if (fieldGroup.type !== 'field-group') {
        return;
    }
    const linkedSection = {
        type: section.type,
        annotation: section,
        configuration: {},
        children: []
    };
    const linkedFieldGroup = {
        type: fieldGroup.type,
        annotation: fieldGroup,
        configuration: {},
        children: []
    };
    linkedSection.children.push(linkedFieldGroup);
    for (const control of [linkedSection, linkedFieldGroup]) {
        if (control.type === 'header-section') {
            page.sections.push(control);
        }
        page.lookup[control.type] ??= [];
        page.lookup[control.type].push(control);
    }
}
