// @sap-ux/eslint-plugin-fiori-tools/lib/project-context/linker/fe-v2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runFeV2Linker = runFeV2Linker;
const utils_1 = require("../utils");
const annotations_1 = require("./annotations");
const createModeValues = ['creationRows', 'creationRowsHiddenInEditMode', 'newPage'];
const tableTypeValues = ['Table', 'ResponsiveTable', 'AnalyticalTable', 'GridTable'];
const statePreservationModeValues = ['persistence', 'discovery'];
/**
 * Creates table configuration object
 *
 * @param pathToPage
 * @param createMode
 * @param tableType
 * @param copy
 */
function createTableConfiguration(pathToPage, createMode, tableType, copy) {
    return {
        createMode: {
            values: createModeValues,
            configurationPath: [...pathToPage, 'component', 'settings', 'tableSettings', 'createMode'],
            valueInFile: createMode
        },
        tableType: {
            values: tableTypeValues,
            configurationPath: [...pathToPage, 'component', 'settings', 'tableSettings', 'type'],
            valueInFile: tableType
        },
        copy: {
            values: [true, false],
            configurationPath: [...pathToPage, 'component', 'settings', 'tableSettings', 'copy'],
            valueInFile: copy
        }
    };
}
/**
 * Creates section table configuration object
 *
 * @param pathToPage
 * @param sectionKey
 * @param createMode
 * @param tableType
 * @param copy
 */
function createSectionTableConfiguration(pathToPage, sectionKey, createMode, tableType, copy) {
    return {
        createMode: {
            values: createModeValues,
            configurationPath: [...pathToPage, 'component', 'settings', 'sections', sectionKey, 'createMode'],
            valueInFile: createMode
        },
        tableType: {
            values: tableTypeValues,
            configurationPath: [
                ...pathToPage,
                'component',
                'settings',
                'sections',
                sectionKey,
                'tableSettings',
                'type'
            ],
            valueInFile: tableType
        },
        copy: {
            values: [true, false],
            configurationPath: [
                ...pathToPage,
                'component',
                'settings',
                'sections',
                sectionKey,
                'tableSettings',
                'copy'
            ],
            valueInFile: copy
        }
    };
}
/**
 * Finds section settings from configuration
 *
 * @param configuration
 */
function findSectionSettings(configuration) {
    let sectionEntityKey = '';
    let createMode;
    let tableType;
    let copy;
    for (const [key, value] of Object.entries(configuration.component?.settings?.sections ?? {})) {
        if (value.createMode !== undefined) {
            sectionEntityKey = key;
            createMode = value.createMode;
        }
        if (value.tableSettings?.type !== undefined) {
            sectionEntityKey = key;
            tableType = value.tableSettings.type;
        }
        if (value.tableSettings?.copy !== undefined) {
            sectionEntityKey = key;
            copy = value.tableSettings.copy;
        }
    }
    return { sectionKey: sectionEntityKey, createMode, tableType, copy };
}
/**
 * Creates linked table for a section
 *
 * @param table
 * @param pathToPage
 * @param sectionSettings
 * @param sectionSettings.sectionKey
 * @param sectionSettings.createMode
 * @param sectionSettings.tableType
 * @param sectionSettings.copy
 */
function createLinkedTableForSection(table, pathToPage, sectionSettings) {
    return {
        type: table.type,
        annotation: table,
        configuration: createSectionTableConfiguration(pathToPage, sectionSettings.sectionKey, sectionSettings.createMode, sectionSettings.tableType, sectionSettings.copy),
        children: []
    };
}
/**
 * Gets entity set and entity type from service
 *
 * @param service
 * @param entitySetName
 */
function getEntityData(service, entitySetName) {
    const entity = service.index.entitySets[entitySetName];
    const entityType = entity?.structuredType;
    return { entity, entityType };
}
/**
 * Creates page configuration
 *
 * @param path
 * @param name
 * @param createMode
 * @param condensedTableLayout
 */
function createPageConfiguration(path, name, createMode, condensedTableLayout) {
    return {
        createMode: {
            values: createModeValues,
            configurationPath: [...path, name, 'component', 'settings', 'createMode'],
            valueInFile: createMode
        },
        condensedTableLayout: {
            values: [true, false],
            configurationPath: [...path, name, 'component', 'settings', 'condensedTableLayout'],
            valueInFile: condensedTableLayout
        }
    };
}
/**
 * Links Fiori Elements v2 application structure.
 * Processes application settings and pages to create a linked model.
 *
 * @param context - Linker context containing parsed application data
 * @returns Linked FE v2 application structure
 */
function runFeV2Linker(context) {
    const manifest = context.app.manifestObject;
    const config = manifest['sap.ui.generic.app'];
    const linkedApp = linkApplicationSettings(context);
    const service = (0, utils_1.getParsedServiceByName)(context.app);
    if (!service) {
        return linkedApp;
    }
    if (config) {
        for (const [name, target] of Object.entries(config?.pages ?? {})) {
            linkPage(context, service, linkedApp, ['sap.ui.generic.app', 'pages'], name, target);
        }
    }
    return linkedApp;
}
/**
 * Links a list report page
 *
 * @param context
 * @param service
 * @param linkedApp
 * @param path
 * @param name
 * @param target
 * @param componentName
 */
function linkListReportPage(context, service, linkedApp, path, name, target, componentName) {
    const entitySetName = target.entitySet;
    if (!entitySetName) {
        return;
    }
    const { entity, entityType } = getEntityData(service, entitySetName);
    if (!entity || !entityType) {
        return;
    }
    const mainService = (0, utils_1.getParsedServiceByName)(context.app);
    if (!mainService) {
        return;
    }
    const table = (0, annotations_1.collectTables)('v2', entityType, mainService);
    const createMode = target.component?.settings?.createMode;
    const condensedTableLayout = target.component?.settings?.condensedTableLayout;
    const page = {
        type: 'list-report-page',
        targetName: name,
        componentName,
        configuration: createPageConfiguration(path, name, createMode, condensedTableLayout),
        entitySetName,
        entity,
        tables: [],
        lookup: {}
    };
    linkListReportTable(page, [...path, name], table, target);
    linkedApp.pages.push(page);
}
/**
 * Links an object page
 *
 * @param context
 * @param service
 * @param linkedApp
 * @param path
 * @param name
 * @param target
 */
function linkObjectPagePage(context, service, linkedApp, path, name, target) {
    const entitySetName = target.entitySet;
    if (!entitySetName) {
        return;
    }
    const { entity, entityType } = getEntityData(service, entitySetName);
    if (!entity || !entityType) {
        return;
    }
    const mainService = (0, utils_1.getParsedServiceByName)(context.app);
    if (!mainService) {
        return;
    }
    const sections = (0, annotations_1.collectSections)('v2', entityType, mainService);
    const createMode = target.component?.settings?.createMode;
    const page = {
        type: 'object-page',
        targetName: name,
        componentName: 'sap.suite.ui.generic.template.ObjectPage',
        configuration: createPageConfiguration(path, name, createMode, undefined),
        entitySetName,
        entity,
        sections: [],
        lookup: {}
    };
    linkObjectPageSections(page, [...path, name], entity, mainService, sections.filter((section) => section.type === 'table-section'), target);
    for (const section of sections.filter((section) => section.type === 'header-section')) {
        (0, annotations_1.collectHeaderSections)(section, page);
    }
    linkedApp.pages.push(page);
}
/**
 * Links a page based on component type
 *
 * @param context
 * @param service
 * @param linkedApp
 * @param path
 * @param name
 * @param target
 */
function linkPage(context, service, linkedApp, path, name, target) {
    const componentName = target?.component?.name;
    if (componentName === 'sap.suite.ui.generic.template.ListReport' ||
        componentName === 'sap.suite.ui.generic.template.AnalyticalListPage') {
        linkListReportPage(context, service, linkedApp, path, name, target, componentName);
    }
    else if (componentName === 'sap.suite.ui.generic.template.ObjectPage') {
        linkObjectPagePage(context, service, linkedApp, path, name, target);
    }
    const pages = target.pages ?? {};
    for (const [key, child] of Object.entries(pages)) {
        linkPage(context, service, linkedApp, [...path, name, 'pages'], key, child);
    }
}
/**
 * Links tables in a List Report page with their configuration.
 * Creates linked table structures from annotation nodes and manifest settings.
 *
 * @param page - FE v2 List Report page
 * @param pathToPage - Path to page in manifest structure
 * @param tables - Array of table annotation nodes
 * @param configuration - Manifest page settings containing table configuration
 */
function linkListReportTable(page, pathToPage, tables, configuration) {
    const controls = {};
    for (const table of tables) {
        const configurationKey = (0, annotations_1.getConfigurationKey)(table.annotationPath);
        const tableSettingsConfig = configuration.component?.settings?.tableSettings ?? {};
        const createMode = tableSettingsConfig.createMode;
        const tableType = tableSettingsConfig.type;
        const copy = tableSettingsConfig.copy;
        const linkedTable = {
            type: table.type,
            annotation: table,
            configuration: createTableConfiguration(pathToPage, createMode, tableType, copy),
            children: []
        };
        controls[`${linkedTable.type}|${configurationKey}`] = linkedTable;
    }
    const configurations = configuration.component?.settings?.sections ?? {};
    for (const [sectionKey, sectionConfig] of Object.entries(configurations)) {
        const tableControl = controls[`table|${sectionKey}`];
        const createMode = sectionConfig.createMode;
        const tableType = sectionConfig.tableSettings?.type;
        const copy = sectionConfig.tableSettings?.copy;
        if (!tableControl) {
            const orphanedSection = {
                type: 'orphan-table',
                configuration: createSectionTableConfiguration(pathToPage, sectionKey, createMode, tableType, copy)
            };
            controls[`${orphanedSection.type}|${sectionKey}`] = orphanedSection;
        }
    }
    for (const control of Object.values(controls)) {
        page.lookup[control.type] ??= [];
        page.lookup[control.type].push(control);
    }
}
/**
 * Links object page sections with their tables and configurations for Fiori Elements V2.
 *
 * @param page - The object page being linked
 * @param pathToPage - Configuration path segments to the page
 * @param entity - The metadata element representing the entity
 * @param service - The parsed OData service
 * @param sections - Array of table section nodes to link
 * @param configuration - Manifest page settings
 */
function linkObjectPageSections(page, pathToPage, entity, service, sections, configuration) {
    const controls = {};
    for (const section of sections) {
        if (section.type !== 'table-section') {
            continue;
        }
        const table = section.children[0];
        if (table.type !== 'table') {
            continue;
        }
        const configurationKey = (0, annotations_1.getConfigurationKey)(table.annotationPath);
        const linkedSection = {
            type: section.type,
            annotation: section,
            configuration: {},
            children: []
        };
        controls[`${section.type}|${configurationKey}`] = linkedSection;
        const sectionSettings = findSectionSettings(configuration);
        const linkedTable = createLinkedTableForSection(table, pathToPage, sectionSettings);
        linkedSection.children.push(linkedTable);
        controls[`${linkedTable.type}|${configurationKey}`] = linkedTable;
    }
    const configurations = configuration.component?.settings?.sections ?? {};
    for (const [sectionKey, sectionConfig] of Object.entries(configurations)) {
        const sectionControl = controls[`table-section|${sectionKey}`];
        if (!sectionControl) {
            const createMode = sectionConfig.createMode;
            const tableType = sectionConfig.tableSettings?.type;
            const copy = sectionConfig.tableSettings?.copy;
            const orphanedSection = {
                type: 'orphan-section',
                configuration: createSectionTableConfiguration(pathToPage, sectionKey, createMode, tableType, copy)
            };
            controls[`${orphanedSection.type}|${sectionKey}|`] = orphanedSection;
        }
    }
    for (const control of Object.values(controls)) {
        if (control.type === 'table-section') {
            page.sections.push(control);
        }
        page.lookup[control.type] ??= [];
        page.lookup[control.type].push(control);
    }
}
/**
 * Links application-level settings from manifest configuration for Fiori Elements V2.
 *
 * @param context - Linker context containing parsed application data
 * @returns A linked Fiori Elements V2 application object
 */
function linkApplicationSettings(context) {
    const config = context.app.manifestObject['sap.ui.generic.app'] ?? {};
    const createMode = config.settings?.tableSettings?.createMode;
    const statePreservationMode = config.settings?.statePreservationMode;
    const twoColumnLayoutValue = config.settings?.flexibleColumnLayout?.defaultTwoColumnLayoutType;
    const threeColumnLayoutValue = config.settings?.flexibleColumnLayout?.defaultThreeColumnLayoutType;
    const tableColumnVerticalAlignmentValue = config.settings?.tableColumnVerticalAlignment;
    const linkedApp = {
        type: 'fe-v2',
        pages: [],
        configuration: {
            createMode: {
                values: createModeValues,
                configurationPath: ['sap.ui.generic.app', 'settings', 'tableSettings', 'createMode'],
                valueInFile: createMode
            },
            statePreservationMode: {
                values: statePreservationModeValues,
                configurationPath: ['sap.ui.generic.app', 'settings', 'statePreservationMode'],
                valueInFile: statePreservationMode
            },
            flexibleColumnLayout: {
                defaultTwoColumnLayoutType: {
                    values: ['TwoColumnsBeginExpanded', 'TwoColumnsMidExpanded'],
                    configurationPath: ['sap.ui.generic.app', 'flexibleColumnLayout', 'defaultTwoColumnLayoutType'],
                    valueInFile: twoColumnLayoutValue
                },
                defaultThreeColumnLayoutType: {
                    values: ['ThreeColumnsMidExpanded', 'ThreeColumnsEndExpanded'],
                    configurationPath: ['sap.ui.generic.app', 'flexibleColumnLayout', 'defaultThreeColumnLayoutType'],
                    valueInFile: threeColumnLayoutValue
                }
            },
            tableColumnVerticalAlignment: {
                values: ['Bottom', 'Middle', 'Top'],
                configurationPath: ['sap.ui.generic.app', 'settings', 'tableColumnVerticalAlignment'],
                valueInFile: tableColumnVerticalAlignmentValue
            }
        }
    };
    return linkedApp;
}
