// @sap/cds-dk/lib/init/template/postgres/index.js
const cds = require('../../../cds')
const { read, exists } = cds.utils
const { readProject } = require('../../projectReader')
const { merge, sort } = require('../../merge')
const mvn = require('../../mvn')
const { srv4, postgres, postgresDeployer } = require('../../registries/mta')
const KymaTemplate = require('../kyma')
const { execSync } = require('child_process')

module.exports = class PostgresTemplate extends require('../../plugin') {

  static help() {
    return 'database support for PostgreSQL'
  }

  static hasInProduction(env) {
    return env.requires?.db?.kind === 'postgres' || readProject().dependencies?.['@cap-js/postgres']
  }

  async run() {
    const project = readProject()
    const { configFile, isJava } = project
    project.shortcut = !exists(configFile) || typeof (await read(configFile)).cds?.requires?.db !== 'object'
    await merge(__dirname, 'files/package.json.hbs').into(configFile, { with: project, forceOverwrite: true }) // REVISIT: forceOverwrite: true as default?
    if (isJava) await mvn.add('postgresql')

    await sort('package.json', 'dependencies')
  }

  async combine() {
    const project = readProject()
    const { addMta, srvPath, hasMultitenancy, addKyma, hasKyma, addMultitenancy } = project

    if (addMta) {
      const srv = srv4(srvPath)
      await merge(__dirname, 'files/mta.yml.hbs').into('mta.yaml', {
        project,
        additions: [srv, postgres, postgresDeployer],
        relationships: [{
            insert: [postgres, 'name'],
            into: [srv, 'requires', 'name']
        }, {
          insert: [postgres, 'name'],
          into: [postgresDeployer, 'requires', 'name']
        }]
      })
    }

    if (!hasMultitenancy && addKyma) {
      const clusterIPGuess = execSync(`kubectl --namespace kyma-system get configmap kyma-info -o json | jq -r '.data["cloud.natGatewayIps"]'`, { stdio: "pipe" }).toString()
      project['KYMA_GATEWAY_IP'] = clusterIPGuess.replace(/\n/g, '')
      await merge(__dirname, 'files', 'values.yaml.hbs').into('chart/values.yaml', { with: project })
      await KymaTemplate.mergeDependency('service-instance', 'postgres')
      await KymaTemplate.mergeDependency('content-deployment', 'postgres-deployer')
    }

    if (addMultitenancy && hasKyma) {
      await merge('chart/values.yaml')
        .remove(['postgres-deployer', 'srv.bindings.db', 'postgres'])

      await merge('chart/Chart.yaml')
        .remove([
            { dependencies: { alias: 'postgres' } },
            { dependencies: { alias: 'postgres-deployer' } }
        ])
    }

     if (addKyma && !hasMultitenancy) {
      const additions = [{
          in: 'modules',
          where: { name: `${project.appName}-postgres-deployer` }
      }]
      await merge(__dirname, '/files/containerize.yaml.hbs').into('containerize.yaml', { with: project, additions })
    }
  }
}
