// @sap-ux/eslint-plugin-fiori-tools/lib/language/xml/source-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FioriXMLSourceCode = exports.visitorKeys = void 0;
exports.isNode = isNode;
const plugin_kit_1 = require("@eslint/plugin-kit");
const traversal_step_1 = require("./traversal-step");
/**
 * Visitor keys mapping XML node types to their traversable child properties.
 * Used to navigate the XML AST during traversal.
 */
exports.visitorKeys = {
    XMLDocument: ['prolog', 'rootElement'],
    XMLProlog: ['attributes'],
    XMLElement: ['attributes', 'subElements', 'textContents'],
    XMLPrologAttribute: [],
    XMLAttribute: [],
    XMLTextContent: []
};
/**
 * XML Source Code representation for Fiori annotation files.
 * Extends TextSourceCodeBase to provide XML-specific traversal and analysis.
 */
class FioriXMLSourceCode extends plugin_kit_1.TextSourceCodeBase {
    projectContext;
    /**
     * The AST of the source code.
     */
    ast;
    /**
     * Creates an instance of FioriXMLSourceCode.
     *
     * @param root0 - Object containing text, ast, and projectContext.
     * @param root0.text - The text content of the file.
     * @param root0.ast - The XML AST.
     * @param root0.projectContext - The project context associated with the given file.
     */
    constructor({ text, ast, projectContext }) {
        super({ text, ast });
        this.ast = ast;
        this.projectContext = projectContext;
    }
    /**
     * Returns the parent of the given node.
     *
     * @param node - The node to get the parent of.
     * @returns The parent of the node.
     * @throws {Error} If the method is not implemented in the subclass.
     */
    getParent(node) {
        if (isNode(node)) {
            if (node.type === 'XMLDocument') {
                return undefined;
            }
            return node.parent;
        }
        else {
            return undefined;
        }
    }
    /**
     * Traverse the source code and return the steps that were taken.
     *
     * @returns The steps that were taken while traversing the source code.
     */
    traverse() {
        const steps = [];
        const visit = (node, parent) => {
            steps.push(new traversal_step_1.XMLVisitNodeStep({
                target: node,
                phase: traversal_step_1.STEP_PHASE.ENTER,
                args: [node, parent]
            }));
            for (const key of exports.visitorKeys[node.type] ?? []) {
                // @ts-expect-error 7053
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                const child = node[key];
                if (child) {
                    if (Array.isArray(child)) {
                        child.forEach((grandchild) => {
                            visit(grandchild, node);
                        });
                    }
                    else {
                        visit(child, node);
                    }
                }
            }
            steps.push(new traversal_step_1.XMLVisitNodeStep({
                target: node,
                phase: traversal_step_1.STEP_PHASE.EXIT,
                args: [node, parent]
            }));
        };
        visit(this.ast);
        return steps;
    }
    /**
     * Get location of node or token.
     *
     * @param node - Node or token
     * @returns Location
     */
    getLoc(node) {
        if (isNode(node)) {
            return {
                start: {
                    line: node.position.startLine,
                    column: node.position.startColumn
                },
                end: {
                    line: node.position.endLine,
                    column: node.position.endColumn
                }
            };
        }
        return {
            start: {
                line: node.startLine,
                column: node.startColumn
            },
            end: {
                line: node.endLine,
                column: node.endColumn
            }
        };
    }
}
exports.FioriXMLSourceCode = FioriXMLSourceCode;
/**
 * Checks if the given object is an XML AST node.
 *
 * @param node - The object to check.
 * @returns True if the object is an XML AST node, false otherwise.
 */
function isNode(node) {
    return node.position !== undefined;
}
