// @sap-ux/eslint-plugin-fiori-tools/lib/language/diagnostics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STRICT_UOM_FILTERING = exports.CONDENSED_TABLE_LAYOUT = exports.NO_DATA_FIELD_INTENT_BASED_NAVIGATION = exports.TABLE_COLUMN_VERTICAL_ALIGNMENT = exports.TABLE_PERSONALIZATION = exports.STATE_PRESERVATION_MODE = exports.CREATION_MODE_FOR_TABLE = exports.ENABLE_PASTE = exports.ENABLE_EXPORT = exports.COPY_TO_CLIPBOARD = exports.FLEX_ENABLED = exports.ANCHOR_BAR_VISIBLE = exports.WIDTH_INCLUDING_COLUMN_HEADER_RULE_TYPE = void 0;
exports.WIDTH_INCLUDING_COLUMN_HEADER_RULE_TYPE = 'sap-width-including-column-header';
exports.ANCHOR_BAR_VISIBLE = 'sap-anchor-bar-visible';
exports.FLEX_ENABLED = 'sap-flex-enabled';
exports.COPY_TO_CLIPBOARD = 'sap-copy-to-clipboard';
exports.ENABLE_EXPORT = 'sap-enable-export';
exports.ENABLE_PASTE = 'sap-enable-paste';
exports.CREATION_MODE_FOR_TABLE = 'sap-creation-mode-for-table';
exports.STATE_PRESERVATION_MODE = 'sap-state-preservation-mode';
exports.TABLE_PERSONALIZATION = 'sap-table-personalization';
exports.TABLE_COLUMN_VERTICAL_ALIGNMENT = 'sap-table-column-vertical-alignment';
exports.NO_DATA_FIELD_INTENT_BASED_NAVIGATION = 'sap-no-data-field-intent-based-navigation';
exports.CONDENSED_TABLE_LAYOUT = 'sap-condensed-table-layout';
exports.STRICT_UOM_FILTERING = 'sap-strict-uom-filtering';
