// @sap-ux/eslint-plugin-fiori-tools/lib/language/fiori-language.js
"use strict";
// provides a "language" according to the eslint documentation (see https://eslint.org/docs/latest/extend/languages) to deal with fiori elements apps
// in contrast to usual eslint languages, this one does not deal only with one file but with a set of files belonging to one app (e.g. manifest.json, annotations, ...)
Object.defineProperty(exports, "__esModule", { value: true });
exports.FioriLanguage = void 0;
const node_url_1 = require("node:url");
const momoa_1 = require("@humanwhocodes/momoa");
const parser_1 = require("@xml-tools/parser");
const ast_1 = require("@xml-tools/ast");
const odata_annotation_core_1 = require("@sap-ux/odata-annotation-core");
const project_access_1 = require("@sap-ux/project-access");
const source_code_1 = require("./json/source-code");
const source_code_2 = require("./xml/source-code");
const project_context_1 = require("../project-context/project-context");
const source_code_3 = require("./annotations/source-code");
const diagnostic_cache_1 = require("./diagnostic-cache");
/**
 * Language implementation for Fiori Elements applications.
 * Provides parsing and validation for manifest.json, XML annotations, and CDS annotation files.
 * Unlike typical ESLint languages, this operates on a set of related files that comprise a Fiori app.
 */
class FioriLanguage {
    fileType = 'text';
    lineStart = 1;
    columnStart = 1;
    nodeTypeKey = 'type';
    /**
     * Validates language options for the Fiori language.
     * Currently no validation is performed.
     *
     * @param _languageOptions - Language options to validate
     */
    validateLanguageOptions(_languageOptions) { } // NOSONAR - Empty method required by ESLint Language interface
    visitorKeys = { ...source_code_2.visitorKeys, ...Object.fromEntries(momoa_1.visitorKeys), ...source_code_3.visitorKeys };
    /**
     * Parses a Fiori file (manifest.json, XML annotations, or CDS annotations).
     * Determines file type and creates appropriate AST representation.
     *
     * @param file - File to parse
     * @param _context - Language context (unused)
     * @returns Parse result containing AST and project context
     */
    parse(file, _context) {
        const text = typeof file.body === 'string' ? file.body : new TextDecoder().decode(file.body);
        const path = (0, project_access_1.normalizePath)(file.path);
        const uri = (0, node_url_1.pathToFileURL)(path).toString();
        // currently this does not work well in ESLint extension,
        // because it triggers this on window focus change and cache is cleared even though file hasn't change
        // similar problem is also with project context update in general
        diagnostic_cache_1.DiagnosticCache.clear(uri);
        const projectContext = project_context_1.ProjectContext.updateFile(uri, text);
        const document = projectContext.index.documents[uri];
        if (!document) {
            if (path.endsWith('.xml')) {
                try {
                    const { cst, tokenVector } = (0, parser_1.parse)(text);
                    const ast = (0, ast_1.buildAst)(cst, tokenVector);
                    return {
                        ok: true,
                        ast: {
                            uri,
                            context: projectContext,
                            document: {
                                type: 'xml',
                                root: ast
                            }
                        }
                    };
                }
                catch (e) {
                    return {
                        ok: false,
                        errors: [
                            {
                                line: 0,
                                column: 0,
                                message: `Failed to parse XML file ${path}: ${e.message}`
                            }
                        ]
                    };
                }
            }
            if (path.endsWith('.json')) {
                return {
                    ok: true,
                    ast: {
                        uri,
                        context: projectContext,
                        document: {
                            type: 'json',
                            root: (0, momoa_1.parse)(text, {
                                mode: 'json',
                                ranges: true,
                                tokens: true,
                                allowTrailingCommas: false
                            })
                        }
                    }
                };
            }
            if (path.endsWith('.cds')) {
                // NOSONAR - TODO: add warning for orphaned cds files
                // we're not checking files that are not part of a fiori app, but we don't want to throw either?
                // other option is to let user add such files to ignore list in eslint config
                return {
                    ok: true,
                    ast: {
                        uri,
                        context: projectContext,
                        document: {
                            type: 'annotation',
                            root: {
                                type: 'annotation-file',
                                references: [],
                                targets: [],
                                uri
                            }
                        }
                    }
                };
            }
            return {
                ok: false,
                errors: [
                    {
                        line: 0,
                        column: 0,
                        message: `File ${path} is not part of a Fiori project or could not be indexed.`
                    }
                ]
            };
        }
        if (document.type === 'Document') {
            return {
                ok: true,
                ast: {
                    uri,
                    context: projectContext,
                    document: {
                        type: 'json',
                        root: document
                    }
                }
            };
        }
        else if (document.type === 'XMLDocument') {
            return {
                ok: true,
                ast: {
                    uri,
                    context: projectContext,
                    document: {
                        type: 'xml',
                        root: document
                    }
                }
            };
        }
        else if (document.type === odata_annotation_core_1.ANNOTATION_FILE_TYPE) {
            return {
                ok: true,
                ast: {
                    uri,
                    context: projectContext,
                    document: {
                        type: 'annotation',
                        root: document
                    }
                }
            };
        }
        throw new Error('Unsupported document type');
    }
    /**
     * Creates SourceCode object that ESLint uses to work with a file.
     *
     * @param file - The file to create SourceCode for.
     * @param parseResult - The result of parsing the file.
     * @returns A SourceCode instance.
     */
    createSourceCode(file, parseResult) {
        const text = typeof file.body === 'string' ? file.body : new TextDecoder().decode(file.body);
        const document = parseResult.ast.document;
        if (document.type === 'json') {
            // Return a minimal SourceCode object as a placeholder
            return new source_code_1.FioriJSONSourceCode({
                text,
                ast: document.root,
                projectContext: parseResult.ast.context,
                uri: parseResult.ast.uri
            });
        }
        else if (document.type === 'xml') {
            return new source_code_2.FioriXMLSourceCode({
                text,
                ast: document.root,
                projectContext: parseResult.ast.context
            });
        }
        else if (document.type === 'annotation') {
            return new source_code_3.FioriAnnotationSourceCode({
                text,
                ast: document.root,
                projectContext: parseResult.ast.context
            });
        }
        throw new Error('Unsupported parse result AST type');
    }
    defaultLanguageOptions;
}
exports.FioriLanguage = FioriLanguage;
