// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-localstorage.js
"use strict";
/**
 * @file detects usage of localstaorage
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            localStorageUsage: 'Local storage must not be used in a Fiori application'
        },
        schema: []
    },
    create(context) {
        const FORBIDDEN_STORAGE_OBJECT = [];
        const storageHelpers = (0, helpers_1.createStorageRuleHelpers)('localStorage', 'localStorageUsage');
        return {
            'VariableDeclarator'(node) {
                storageHelpers.processVariableDeclarator(node, FORBIDDEN_STORAGE_OBJECT);
            },
            'MemberExpression'(node) {
                storageHelpers.checkMemberExpression(node, FORBIDDEN_STORAGE_OBJECT, context);
            }
        };
    }
};
exports.default = rule;
