// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-flex-enabled.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const rule_factory_1 = require("../language/rule-factory");
const diagnostics_1 = require("../language/diagnostics");
const version_1 = require("../utils/version");
const rule_fixer_1 = require("../language/rule-fixer");
const rule = (0, rule_factory_1.createFioriRule)({
    ruleId: diagnostics_1.FLEX_ENABLED,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'The "flexEnabled" flag in the manifest.json file indicates that your application is enabled for UI adaptation.',
            url: 'https://ui5.sap.com/sdk/#/topic/f1430c0337534d469da3a56307ff76af'
        },
        messages: {
            [diagnostics_1.FLEX_ENABLED]: '"flexEnabled" must be set to "true" to enable UI adaptation features.'
        },
        fixable: 'code'
    },
    check(context) {
        const problems = [];
        for (const [, app] of Object.entries(context.sourceCode.projectContext.index.apps)) {
            if (!app.manifest.minUI5Version ||
                (0, version_1.isLowerThanMinimalUi5Version)(app.manifest.minUI5Version, { major: 1, minor: 56 })) {
                continue;
            }
            if (!app.manifest.flexEnabled) {
                problems.push({
                    type: diagnostics_1.FLEX_ENABLED,
                    manifest: {
                        uri: app.manifest.manifestUri,
                        object: app.manifestObject,
                        propertyPath: ['sap.ui5', 'flexEnabled']
                    }
                });
            }
        }
        return problems;
    },
    createJsonVisitorHandler: (context, diagnostic, paths) => function report(node) {
        context.report({
            node,
            messageId: diagnostics_1.FLEX_ENABLED,
            fix: (0, rule_fixer_1.createJsonFixer)({
                value: true,
                context,
                deepestPathResult: paths,
                node
            })
        });
    }
});
exports.default = rule;
