// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-history-manipulation.js
"use strict";
/**
 * @file Detect some warning for usages of (window.)document APIs
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Helper Functions
// ------------------------------------------------------------------------------
/**
 * Check if a node represents a condition statement.
 *
 * @param node The AST node to check
 * @returns True if the node represents a condition statement
 */
function isCondition(node) {
    return (0, helpers_1.isType)(node, 'IfStatement') || (0, helpers_1.isType)(node, 'ConditionalExpression');
}
/**
 * Check if a node represents a unary expression.
 *
 * @param node The AST node to check
 * @returns True if the node represents a unary expression
 */
function isUnary(node) {
    return (0, helpers_1.isType)(node, 'UnaryExpression');
}
/**
 * Check if a node is within a conditional statement up to a maximum depth.
 *
 * @param node The AST node to check
 * @param maxDepth Maximum depth to search for conditions
 * @returns True if the node is within a conditional statement
 */
function isInCondition(node, maxDepth) {
    // we check the depth here because the call might be nested in a block statement and in an expression statement (http://jointjs.com/demos/javascript-ast)
    // (true?history.back():''); || if(true) history.back(); || if(true){history.back();} || if(true){}else{history.back();}
    if (maxDepth > 0) {
        const parent = node.parent;
        return isCondition(parent) || isInCondition(parent, maxDepth - 1);
    }
    return false;
}
/**
 * Check if a node represents the value -1.
 *
 * @param node The AST node to check
 * @returns True if the node represents the value -1
 */
function isMinusOne(node) {
    return isUnary(node) && node.operator === '-' && (0, helpers_1.isLiteral)(node.argument) && node.argument.value === 1;
}
/**
 * Check if a history manipulation call is valid.
 *
 * @param node The call expression node to validate
 * @returns True if the history manipulation call is valid
 */
function isValid(node) {
    switch (node.callee.property.name) {
        case 'forward':
            return false;
        case 'back':
            return isInCondition(node, 3);
        case 'go': {
            const args = node.arguments;
            return args.length === 1 && isMinusOne(args[0]) && isInCondition(node, 3);
        }
        default:
    }
    return true;
}
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            historyManipulation: 'Direct history manipulation, does not work with deep links, use router and navigation events instead'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        const HISTORY_OBJECTS = [];
        // Initialize factory functions
        const isWindowObject = (0, helpers_1.createIsWindowObject)(WINDOW_OBJECTS);
        const rememberWindow = (0, helpers_1.createRememberWindow)(WINDOW_OBJECTS, isWindowObject);
        const isHistory = (0, helpers_1.createIsHistory)(isWindowObject);
        const isHistoryObject = (0, helpers_1.createIsHistoryObject)(HISTORY_OBJECTS, isHistory);
        const rememberHistory = (0, helpers_1.createRememberHistory)(HISTORY_OBJECTS, isHistoryObject);
        // --------------------------------------------------------------------------
        // Helper Functions
        // --------------------------------------------------------------------------
        /**
         * Check if a call expression is interesting for history manipulation detection.
         *
         * @param node The call expression node to check
         * @returns True if the call expression is interesting for history manipulation detection
         */
        function isInteresting(node) {
            // check if callee is ref to history.back / .go / .forward
            if (node &&
                (0, helpers_1.isMember)(node.callee) &&
                isHistoryObject(node.callee.object) &&
                (0, helpers_1.isIdentifier)(node.callee.property)) {
                return true;
            }
            return false;
        }
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'CallExpression': function (node) {
                if (isInteresting(node) && !isValid(node)) {
                    context.report({ node: node, messageId: 'historyManipulation' });
                }
            },
            'VariableDeclarator': function (node) {
                return rememberWindow(node.id, node.init) || rememberHistory(node.id, node.init);
            },
            'AssignmentExpression': function (node) {
                return rememberWindow(node.left, node.right) || rememberHistory(node.left, node.right);
            }
        };
    }
};
exports.default = rule;
