// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-exec-command.js
"use strict";
/**
 * @file Detect usage of execCommand
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            execCommand: 'Direct DOM Manipulation, better to use jQuery.appendTo if really needed'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        const DOCUMENT_OBJECTS = [];
        // Initialize factory functions
        const isWindowObject = (0, helpers_1.createIsWindowObject)(WINDOW_OBJECTS);
        const rememberWindow = (0, helpers_1.createRememberWindow)(WINDOW_OBJECTS, isWindowObject);
        const isDocument = (0, helpers_1.createIsDocument)(isWindowObject);
        const isDocumentObject = (0, helpers_1.createIsDocumentObject)(DOCUMENT_OBJECTS, isDocument);
        const rememberDocument = (0, helpers_1.createRememberDocument)(DOCUMENT_OBJECTS, isDocumentObject);
        // --------------------------------------------------------------------------
        // Helpers
        // --------------------------------------------------------------------------
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator': function (node) {
                return rememberWindow(node.id, node.init) || rememberDocument(node.id, node.init);
            },
            'AssignmentExpression': function (node) {
                return rememberWindow(node.left, node.right) || rememberDocument(node.left, node.right);
            },
            'MemberExpression': function (node) {
                if (isDocumentObject(node.object) &&
                    (((0, helpers_1.isIdentifier)(node.property) && 'name' in node.property && node.property.name === 'execCommand') ||
                        ((0, helpers_1.isLiteral)(node.property) && 'value' in node.property && node.property.value === 'execCommand'))) {
                    context.report({ node: node, messageId: 'execCommand' });
                }
            }
        };
    }
};
exports.default = rule;
