// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-event-prop.js
"use strict";
/**
 * @file Rule to flag use of a private member from UI5 Event
 */
Object.defineProperty(exports, "__esModule", { value: true });
// THIS RULE IS DEPRECATED --> sap-no-ui5base-prop
// ------------------------------------------------------------------------------
// Helper Functions
// ------------------------------------------------------------------------------
/**
 * Check if an array contains a specific object.
 *
 * @param a The array to search in
 * @param obj The object to search for
 * @returns True if the array contains the object
 */
function contains(a, obj) {
    return a.includes(obj);
}
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            eventProp: 'Direct usage of a private member from  sap.ui.base.Event detected!'
        },
        schema: []
    },
    create(context) {
        const PRIVATE_MEMBERS = ['oSource', 'mParameters', 'sId'];
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'MemberExpression': function (node) {
                if (!node.property || !('name' in node.property)) {
                    return;
                }
                const val = node.property.name;
                if (typeof val === 'string' && contains(PRIVATE_MEMBERS, val)) {
                    context.report({
                        node: node,
                        messageId: 'eventProp'
                    });
                }
            }
        };
    }
};
exports.default = rule;
