// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-dynamic-style-insertion.js
"use strict";
/**
 * @file Detect usage of document.styleSheets
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            dynamicStyleInsertion: 'Dynamic style insertion, use library CSS or lessifier instead'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        const DOCUMENT_OBJECTS = [];
        // Initialize factory functions
        const isWindowObject = (0, helpers_1.createIsWindowObject)(WINDOW_OBJECTS);
        const rememberWindow = (0, helpers_1.createRememberWindow)(WINDOW_OBJECTS, isWindowObject);
        const isDocument = (0, helpers_1.createIsDocument)(isWindowObject);
        const isDocumentObject = (0, helpers_1.createIsDocumentObject)(DOCUMENT_OBJECTS, isDocument);
        const rememberDocument = (0, helpers_1.createRememberDocument)(DOCUMENT_OBJECTS, isDocumentObject);
        // --------------------------------------------------------------------------
        // Helpers
        // --------------------------------------------------------------------------
        /**
         * Check if a node represents an interesting dynamic style insertion.
         *
         * @param node The AST node to check
         * @returns True if the node represents dynamic style insertion, false otherwise
         */
        function isInteresting(node) {
            if ((0, helpers_1.isMember)(node) && (0, helpers_1.isMember)(node.object) && isDocumentObject(node.object.object)) {
                const prop = node.object.property;
                if ((0, helpers_1.isIdentifier)(prop) && prop.name === 'styleSheets') {
                    return true;
                }
                if ((0, helpers_1.isLiteral)(prop) && prop.value === 'styleSheets') {
                    return true;
                }
            }
            return false;
        }
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator': function (node) {
                return rememberWindow(node.id, node.init) || rememberDocument(node.id, node.init);
            },
            'AssignmentExpression': function (node) {
                return rememberWindow(node.left, node.right) || rememberDocument(node.left, node.right);
            },
            'MemberExpression': function (node) {
                if (isInteresting(node)) {
                    context.report({ node: node, messageId: 'dynamicStyleInsertion' });
                }
            }
        };
    }
};
exports.default = rule;
