// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-inner-html-access.js
"use strict";
/**
 * @file Detect the access of the innerHTML property.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
/**
 * Check if a property access is valid (not innerHTML).
 *
 * @param property The property node to validate
 * @returns True if the property access is valid
 */
function isValid(property) {
    // anything is valid, except 'innerHTML'
    if ((0, helpers_1.isIdentifier)(property)) {
        return property.name !== 'innerHTML';
    }
    else if ((0, helpers_1.isLiteral)(property)) {
        return property.value !== 'innerHTML';
    }
    return true;
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            innerHtmlAccess: 'Accessing the inner html is not recommended.'
        },
        schema: []
    },
    create(context) {
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'MemberExpression'(node) {
                if (!isValid(node.property)) {
                    context.report({ node: node, messageId: 'innerHtmlAccess' });
                }
            }
        };
    }
};
exports.default = rule;
