// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-localhost.js
"use strict";
/**
 * @file     Check "sap-no-localhost" should detect the usage of "localhost".
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
/**
 * Check if a string does not contain a substring.
 *
 * @param string The string to search in
 * @param substring The substring to check for absence
 * @returns True if the string does not contain the substring
 */
function containsNot(string, substring) {
    return !(0, helpers_1.containsString)(string, substring);
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            localhost: "Usage of 'localhost' detected"
        },
        schema: []
    },
    create(context) {
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'Literal': function (node) {
                if ((0, helpers_1.isString)(node.value) &&
                    (0, helpers_1.containsString)(node.value, 'localhost') &&
                    containsNot(node.value, '://localhost/offline/')) {
                    context.report({ node: node, messageId: 'localhost' });
                }
            }
        };
    }
};
exports.default = rule;
