// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-br-on-return.js
"use strict";
/**
 * @file Detect the usage of document.queryCommandSupported with 'insertBrOnReturn' argument
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            insertBrOnReturn: "insertBrOnReturn is not allowed since it is a Mozilla specific method, other browsers don't support that."
        },
        schema: []
    },
    create(context) {
        const createVisitors = (0, helpers_1.createDocumentBasedRuleVisitors)({
            isInteresting: (node, isDocumentObject) => {
                return ((0, helpers_1.isCall)(node.parent) &&
                    isDocumentObject(node.object) &&
                    (((0, helpers_1.isIdentifier)(node.property) &&
                        node.property.name === 'queryCommandSupported') ||
                        ((0, helpers_1.isLiteral)(node.property) && node.property.value === 'queryCommandSupported')));
            },
            isValid: (node) => {
                return (node.parent.arguments.length === 0 ||
                    node.parent.arguments[0].value !== 'insertBrOnReturn');
            },
            messageId: 'insertBrOnReturn'
        });
        return createVisitors(context);
    }
};
exports.default = rule;
