// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-window-alert.js
"use strict";
/**
 * @file Rule to flag use of alert
 */
Object.defineProperty(exports, "__esModule", { value: true });
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            windowAlert: 'A window.alert statement should not be part of the code that is committed to GIT! Use sap.m.MessageBox instead.'
        },
        schema: []
    },
    create(context) {
        return {
            'CallExpression': function (node) {
                if (node.callee.type === 'MemberExpression' && node.callee.property.type === 'Identifier') {
                    if (node.callee.object &&
                        'name' in node.callee.object &&
                        node.callee.object.name === 'window' &&
                        node.callee.property &&
                        'name' in node.callee.property &&
                        node.callee.property.name === 'alert') {
                        context.report({
                            node: node,
                            messageId: 'windowAlert'
                        });
                    }
                }
            }
        };
    }
};
exports.default = rule;
