// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-ui5-legacy-factories.js
"use strict";
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            legacyFactories: 'Make use of sap.ui.define([...], function(...) {...} to load required dependencies. Legacy UI5 factories leading to synchronous loading.'
        },
        schema: []
    },
    create(context) {
        const INTERESTING_METHODS = [
            'jsview',
            'component',
            'view',
            'xmlview',
            'controller',
            'extensionpoint',
            'fragment',
            'getVersionInfo',
            'resources',
            'load'
        ];
        const INTERESTING_METHODS_JQUERY = ['resources'];
        // --------------------------------------------------------------------------
        // Basic Helpers
        // --------------------------------------------------------------------------
        /**
         * Check if callee matches sap.ui.* pattern
         *
         * @param callee
         */
        function isSapUiPattern(callee) {
            return ((0, helpers_1.isMember)(callee) &&
                (0, helpers_1.isIdentifier)(callee.property) &&
                (0, helpers_1.contains)(INTERESTING_METHODS, callee.property.name) &&
                (0, helpers_1.isMember)(callee.object) &&
                callee.object.property.name === 'ui' &&
                (0, helpers_1.isIdentifier)(callee.object.object) &&
                callee.object.object.name === 'sap');
        }
        /**
         * Check if callee matches jQuery.sap.* or $.sap.* pattern
         *
         * @param callee
         */
        function isJQuerySapPattern(callee) {
            return ((0, helpers_1.isMember)(callee) &&
                (0, helpers_1.isIdentifier)(callee.property) &&
                (0, helpers_1.contains)(INTERESTING_METHODS_JQUERY, callee.property.name) &&
                (0, helpers_1.isMember)(callee.object) &&
                callee.object.property.name === 'sap' &&
                (0, helpers_1.isIdentifier)(callee.object.object) &&
                (callee.object.object.name === 'jQuery' || callee.object.object.name === '$'));
        }
        /**
         * Check if callee matches sap.ui.component.* pattern
         *
         * @param callee
         */
        function isSapUiComponentPattern(callee) {
            return ((0, helpers_1.isMember)(callee) &&
                (0, helpers_1.isIdentifier)(callee.property) &&
                (0, helpers_1.contains)(INTERESTING_METHODS, callee.property.name) &&
                (0, helpers_1.isMember)(callee.object) &&
                callee.object.property.name === 'component' &&
                (0, helpers_1.isMember)(callee.object.object) &&
                callee.object.object.property.name === 'ui' &&
                (0, helpers_1.isIdentifier)(callee.object.object.object) &&
                callee.object.object.object.name === 'sap');
        }
        /**
         * Check if a call expression represents an interesting legacy factory usage.
         *
         * @param node The call expression node to check
         * @returns True if the call expression represents an interesting legacy factory usage
         */
        function isInteresting(node) {
            const callee = node.callee;
            return isSapUiPattern(callee) || isJQuerySapPattern(callee) || isSapUiComponentPattern(callee);
        }
        return {
            'CallExpression': function (node) {
                if (isInteresting(node)) {
                    context.report({ node: node, messageId: 'legacyFactories' });
                }
            }
        };
    }
};
exports.default = rule;
