// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-timeout-usage.js
"use strict";
/**
 * @file Detect some warning for usages of (window.)document APIs
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
/**
 * Check if a node represents setTimeout.
 *
 * @param node The AST node to check
 * @returns True if the node represents setTimeout
 */
function isTimeout(node) {
    return (0, helpers_1.isIdentifier)(node) && node.name === 'setTimeout';
}
/**
 * Check if a setTimeout call has valid timeout arguments.
 *
 * @param node The AST node to check
 * @returns True if the setTimeout call has valid timeout arguments
 */
function isValid(node) {
    const args = node.arguments;
    return (args &&
        (args.length === 1 ||
            (args.length > 1 &&
                (args[1].value === 0 ||
                    args[1].value === '0'))));
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            timeoutUsage: 'Timeout with value > 0'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        // --------------------------------------------------------------------------
        // Helpers
        // --------------------------------------------------------------------------
        /**
         * Check if a node represents an interesting setTimeout call to analyze.
         *
         * @param node The AST node to check
         * @returns True if the node represents an interesting setTimeout call
         */
        function isInteresting(node) {
            let obj = node.callee;
            if ((0, helpers_1.isMember)(obj)) {
                const memberObj = obj;
                if ((0, helpers_1.isWindow)(memberObj.object) ||
                    ((0, helpers_1.isIdentifier)(memberObj.object) &&
                        (0, helpers_1.contains)(WINDOW_OBJECTS, memberObj.object.name))) {
                    // is member expression on window, check property
                    obj = memberObj.property;
                }
                else {
                    // no call on window
                    return false;
                }
            }
            // here obj may not be node.callee any more but node.callee.property
            return isTimeout(obj);
        }
        /**
         * Remember window object references for later analysis.
         *
         * @param left The left side of the assignment
         * @param right The right side of the assignment
         */
        function rememberWindow(left, right) {
            if (right && (0, helpers_1.isWindow)(right) && left && (0, helpers_1.isIdentifier)(left)) {
                WINDOW_OBJECTS.push(left.name);
            }
        }
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator': function (node) {
                const declaratorNode = node;
                if (declaratorNode.init) {
                    rememberWindow(declaratorNode.id, declaratorNode.init);
                }
            },
            'AssignmentExpression': function (node) {
                const assignmentNode = node;
                rememberWindow(assignmentNode.left, assignmentNode.right);
            },
            'CallExpression': function (node) {
                if (isInteresting(node) && !isValid(node)) {
                    context.report({ node: node, messageId: 'timeoutUsage' });
                }
            }
        };
    }
};
exports.default = rule;
