// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-navigator.js
"use strict";
/**
 * @file Detect usage of navigator object
 */
Object.defineProperty(exports, "__esModule", { value: true });
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
/**
 * Check if a node is of a specific type.
 *
 * @param node The AST node to check
 * @param type The type to check for
 * @returns True if the node is of the specified type
 */
function isType(node, type) {
    return node?.type === type;
}
/**
 * Check if a node is an Identifier.
 *
 * @param node The AST node to check
 * @returns True if the node is an Identifier
 */
function isIdentifier(node) {
    return isType(node, 'Identifier');
}
/**
 * Check if a node is a MemberExpression.
 *
 * @param node The AST node to check
 * @returns True if the node is a MemberExpression
 */
function isMember(node) {
    return isType(node, 'MemberExpression');
}
/**
 * Check if a node is a CallExpression.
 *
 * @param node The AST node to check
 * @returns True if the node is a CallExpression
 */
function isCall(node) {
    return isType(node, 'CallExpression');
}
/**
 * Get the rightmost method name from a call expression node.
 *
 * @param node The call expression node to analyze
 * @returns The rightmost method name
 */
function getRightestMethodName(node) {
    const callee = node.callee;
    return isMember(callee) ? callee.property.name : callee.name;
}
/**
 * Check if a node represents the global window object.
 *
 * @param node The AST node to check
 * @returns True if the node represents the global window object
 */
function isWindow(node) {
    // true if node is the global variable 'window'
    return !!(isIdentifier(node) && node && 'name' in node && node.name === 'window');
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            navigator: 'navigator usage is forbidden, use sap.ui.Device API instead'
        },
        schema: []
    },
    create(context) {
        const FORBIDDEN_NAVIGATOR_WINDOW = ['javaEnabled'], FORBIDDEN_GLOB_EVENT = [
            'onload',
            'onunload',
            'onabort',
            'onbeforeunload',
            'onerror',
            'onhashchange',
            'onpageshow',
            'onpagehide',
            'onscroll',
            'onblur',
            'onchange',
            'onfocus',
            'onfocusin',
            'onfocusout',
            'oninput',
            'oninvalid',
            'onreset',
            'onsearch',
            'onselect',
            'onsubmit',
            'onresize'
        ];
        const FORBIDDEN_METHODS = FORBIDDEN_NAVIGATOR_WINDOW.concat(FORBIDDEN_GLOB_EVENT);
        FORBIDDEN_METHODS.push('back');
        const WINDOW_OBJECTS = [];
        const NAVIGATOR_OBJECTS = [];
        // --------------------------------------------------------------------------
        // Helpers
        // --------------------------------------------------------------------------
        /**
         * Check if a node represents the window object or a reference to it.
         *
         * @param node The AST node to check
         * @returns True if the node represents the window object or a reference to it
         */
        function isWindowObject(node) {
            // true if node is the global variable 'window' or a reference to it
            return !!(isWindow(node) ||
                (isIdentifier(node) && node && 'name' in node && WINDOW_OBJECTS.includes(node.name)));
        }
        /**
         * Check if a node represents the navigator object.
         *
         * @param node The AST node to check
         * @returns True if the node represents the navigator object
         */
        function isNavigator(node) {
            // true if node id the global variable 'navigator', 'window.navigator' or '<windowReference>.navigator'
            return ((isIdentifier(node) && node && 'name' in node && node.name === 'navigator') ||
                (isMember(node) && isWindowObject(node.object) && isNavigator(node.property)));
        }
        /**
         * Check if a node represents the navigator object or a reference to it.
         *
         * @param node The AST node to check
         * @returns True if the node represents the navigator object or a reference to it
         */
        function isNavigatorObject(node) {
            // true if node is the global variable 'navigator'/'window.navigator' or a reference to it
            return !!(isNavigator(node) ||
                (isIdentifier(node) && node && 'name' in node && NAVIGATOR_OBJECTS.includes(node.name)));
        }
        /**
         * Remember window object assignments for tracking references.
         *
         * @param left The left-hand side of the assignment
         * @param right The right-hand side of the assignment
         * @returns True if the assignment was remembered
         */
        function rememberWindow(left, right) {
            if (isWindowObject(right) && isIdentifier(left) && 'name' in left) {
                WINDOW_OBJECTS.push(left.name);
                return true;
            }
            return false;
        }
        /**
         * Remember navigator object assignments for tracking references.
         *
         * @param left The left-hand side of the assignment
         * @param right The right-hand side of the assignment
         * @returns True if the assignment was remembered
         */
        function rememberNavigator(left, right) {
            if (isNavigatorObject(right) && isIdentifier(left) && 'name' in left) {
                NAVIGATOR_OBJECTS.push(left.name);
                return true;
            }
            return false;
        }
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator': function (node) {
                return rememberWindow(node.id, node.init) || rememberNavigator(node.id, node.init);
            },
            'AssignmentExpression': function (node) {
                return rememberWindow(node.left, node.right) || rememberNavigator(node.left, node.right);
            },
            'MemberExpression': function (node) {
                if (isNavigatorObject(node.object)) {
                    if (isCall(node.parent)) {
                        const methodName = getRightestMethodName(node.parent);
                        if (typeof methodName === 'string' && FORBIDDEN_METHODS.includes(methodName)) {
                            context.report({ node: node, messageId: 'navigator' });
                        }
                    }
                    else {
                        context.report({ node: node, messageId: 'navigator' });
                    }
                }
            }
        };
    }
};
exports.default = rule;
