// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-dom-access.js
"use strict";
/**
 * @file Detect some warning for usages of (window.)document APIs
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            domAccess: 'Direct DOM access, use jQuery selector instead'
        },
        schema: []
    },
    create(context) {
        const FORBIDDEN_DOCUMENT_METHODS = [
            'getElementById',
            'getElementsByClassName',
            'getElementsByName',
            'getElementsByTagName'
        ];
        const createVisitors = (0, helpers_1.createDocumentBasedRuleVisitors)({
            isInteresting: (node, isDocumentObject) => {
                return node && isDocumentObject(node.object);
            },
            isValid: (node) => {
                return !((0, helpers_1.isIdentifier)(node.property) &&
                    (0, helpers_1.contains)(FORBIDDEN_DOCUMENT_METHODS, node.property.name));
            },
            messageId: 'domAccess'
        });
        return createVisitors(context);
    }
};
exports.default = rule;
