// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-strict-uom-filtering.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const diagnostics_1 = require("../language/diagnostics");
const rule_factory_1 = require("../language/rule-factory");
const rule_fixer_1 = require("../language/rule-fixer");
const version_1 = require("../utils/version");
const rule = (0, rule_factory_1.createFioriRule)({
    ruleId: diagnostics_1.STRICT_UOM_FILTERING,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'The "disableStrictUomFiltering" setting in sap.fe.app should not be set to true, as it disables strict unit-of-measure filtering.',
            url: 'https://github.com/SAP/open-ux-tools/blob/main/packages/eslint-plugin-fiori-tools/docs/rules/sap-strict-uom-filtering.md'
        },
        messages: {
            [diagnostics_1.STRICT_UOM_FILTERING]: 'The "disableStrictUomFiltering" property should not be set to true. Remove this property or set it to false to keep strict unit-of-measure filtering active.'
        },
        fixable: 'code'
    },
    check(context) {
        const problems = [];
        for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
            if (app.type !== 'fe-v4') {
                continue;
            }
            const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
            if (!parsedApp.manifest.minUI5Version ||
                (0, version_1.isLowerThanMinimalUi5Version)(parsedApp.manifest.minUI5Version, { major: 1, minor: 143 })) {
                continue;
            }
            if (app.configuration.disableStrictUomFiltering.valueInFile === true) {
                problems.push({
                    type: diagnostics_1.STRICT_UOM_FILTERING,
                    manifest: {
                        uri: parsedApp.manifest.manifestUri,
                        object: parsedApp.manifestObject,
                        propertyPath: app.configuration.disableStrictUomFiltering.configurationPath
                    }
                });
            }
        }
        return problems;
    },
    createJsonVisitorHandler: (context, diagnostic, deepestPathResult) => function report(node) {
        context.report({
            node,
            messageId: diagnostics_1.STRICT_UOM_FILTERING,
            fix: (0, rule_fixer_1.createJsonFixer)({
                context,
                deepestPathResult,
                node,
                operation: 'delete'
            })
        });
    }
});
exports.default = rule;
