// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-inner-html-write.js
"use strict";
/**
 * @file Detect the overriding of the innerHTML.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
/**
 * Check if a left-hand side expression is interesting for analysis.
 *
 * @param left The left-hand side expression to check
 * @returns True if the expression is interesting for analysis
 */
function isInteresting(left) {
    return (0, helpers_1.isMember)(left);
}
/**
 * Check if a property access is valid (not innerHTML).
 *
 * @param property The property node to validate
 * @returns True if the property access is valid
 */
function isValid(property) {
    // anything is valid, except 'innerHTML'
    if ((0, helpers_1.isIdentifier)(property)) {
        return property.name !== 'innerHTML';
    }
    else if ((0, helpers_1.isLiteral)(property)) {
        return property.value !== 'innerHTML';
    }
    return true;
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            innerHtmlWrite: 'Writing to the inner html is not allowed.'
        },
        schema: []
    },
    create(context) {
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'AssignmentExpression'(node) {
                if (isInteresting(node.left) &&
                    'property' in node.left &&
                    !isValid(node.left.property)) {
                    context.report({ node: node, messageId: 'innerHtmlWrite' });
                }
            }
        };
    }
};
exports.default = rule;
