// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-ui5base-prop.js
"use strict";
/**
 * @file Rule to flag use of sap ui5base prop
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            managedObjectProp: 'Property {{property}} is a private member of sap.ui.base.ManagedObject!',
            odataModelProp: 'Property {{property}} is a private member of sap.ui.model.odata.ODataModel or sap.ui.model.odata.v2.ODataModel!',
            eventProviderProp: 'Property {{property}} is a private member of sap.ui.base.EventProvider!',
            eventProp: 'Property {{property}} is a private member of sap.ui.base.Event!'
        },
        schema: []
    },
    create(context) {
        const MANAGED_OBJECT_MEMBERS = [
            'mProperties',
            'mAggregations',
            'mAssociations',
            'mMethods',
            'oParent',
            'aDelegates',
            'aBeforeDelegates',
            'iSuppressInvalidate',
            'oPropagatedProperties',
            'oModels',
            'oBindingContexts',
            'mBindingInfos',
            'sBindingPath',
            'mBindingParameters',
            'mBoundObjects'
        ];
        const EVENT_PROVIDER_MEMBERS = ['mEventRegistry', 'oEventPool'];
        const EVENT_MEMBERS = ['oSource', 'mParameters', 'sId'];
        // oMetadadata has been removed from the of forbidden oDataModel members as there is no API available (June 2015, info by Malte Wedel)
        const ODATA_MODEL_MEMBERS = [
            'oServiceData',
            'bCountSupported',
            'bCache',
            'oRequestQueue',
            'aBatchOperations',
            'oHandler',
            'mSupportedBindingModes',
            'sDefaultBindingMode',
            'bJSON',
            'aPendingRequestHandles',
            'aCallAfterUpdate',
            'mRequests',
            'mDeferredRequests',
            'mChangedEntities',
            'mChangeHandles',
            'mDeferredBatchGroups',
            'mChangeBatchGroups',
            'bTokenHandling',
            'bWithCredentials',
            'bUseBatch',
            'bRefreshAfterChange',
            'sMaxDataServiceVersion',
            'bLoadMetadataAsync',
            'bLoadAnnotationsJoined',
            'sAnnotationURI',
            'sDefaultCountMode',
            'sDefaultOperationMode',
            'oMetadataLoadEvent',
            'oMetadataFailedEvent',
            'sRefreshBatchGroupId',
            'sDefaultChangeBatchGroup',
            /*"oMetadata", */ 'oAnnotations',
            'aUrlParams'
        ];
        const checkMemberExpression = (0, helpers_1.createPropertyChecker)({
            managedObjectProp: MANAGED_OBJECT_MEMBERS,
            eventProviderProp: EVENT_PROVIDER_MEMBERS,
            eventProp: EVENT_MEMBERS,
            odataModelProp: ODATA_MODEL_MEMBERS
        });
        return {
            'MemberExpression'(node) {
                checkMemberExpression(node, context);
            }
        };
    }
};
exports.default = rule;
