// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-ui5-legacy-jquerysap-usage.js
"use strict";
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            legacyJquerysapUsage: 'Legacy jQuery.sap usage is not allowed due to strict Content Security Policy.'
        },
        schema: []
    },
    create(context) {
        const INTERESTING_METHODS_JQUERY = ['require', 'declare'];
        // --------------------------------------------------------------------------
        // Basic Helpers
        // --------------------------------------------------------------------------
        /**
         * Check if a call expression represents an interesting jQuery.sap usage.
         *
         * @param node The call expression node to check
         * @returns True if the call expression represents an interesting jQuery.sap usage
         */
        function isInteresting(node) {
            const callee = node.callee;
            if ((0, helpers_1.isMember)(callee)) {
                if ((0, helpers_1.isIdentifier)(callee.property) && (0, helpers_1.contains)(INTERESTING_METHODS_JQUERY, callee.property.name)) {
                    if ((0, helpers_1.isMember)(callee.object) && callee.object.property.name == 'sap') {
                        if ((0, helpers_1.isIdentifier)(callee.object.object) &&
                            (callee.object.object.name == 'jQuery' || callee.object.object.name == '$')) {
                            return true;
                        }
                    }
                }
            }
            return false;
        }
        return {
            'CallExpression': function (node) {
                if (isInteresting(node)) {
                    context.report({ node: node, messageId: 'legacyJquerysapUsage' });
                }
            }
        };
    }
};
exports.default = rule;
