// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-ui5-prop-warning.js
"use strict";
/**
 * @file Rule to flag use of sap ui5base prop
 */
Object.defineProperty(exports, "__esModule", { value: true });
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
/**
 * Check if an array contains a specific object.
 *
 * @param a The array to search in
 * @param obj The object to search for
 * @returns True if the array contains the object
 */
function contains(a, obj) {
    return a.includes(obj);
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            ui5PropWarning: 'Property {{property}} is a private member of sap.ui.model.odata.v2.ODataModel'
        },
        schema: []
    },
    create(context) {
        // variables should be defined here
        const ODATA_MODEL_V2_MEMBERS = ['oData'];
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'MemberExpression': function (node) {
                if (!node.property || !('name' in node.property)) {
                    return;
                }
                const val = node.property.name;
                if (typeof val === 'string' && contains(ODATA_MODEL_V2_MEMBERS, val)) {
                    context.report({
                        node: node,
                        messageId: 'ui5PropWarning',
                        data: { property: val }
                    });
                }
            }
        };
    }
};
exports.default = rule;
