// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-dom-insertion.js
"use strict";
/**
 * @file Detect direct DOM insertion
 */
Object.defineProperty(exports, "__esModule", { value: true });
// ------------------------------------------------------------------------------
// Helper Functions
// ------------------------------------------------------------------------------
/**
 * Check if a node is of a specific type.
 *
 * @param node The AST node to check
 * @param type The type to check for
 * @returns True if the node is of the specified type
 */
function isType(node, type) {
    return node?.type === type;
}
/**
 * Check if a node is an Identifier.
 *
 * @param node The AST node to check
 * @returns True if the node is an Identifier
 */
function isIdentifier(node) {
    return isType(node, 'Identifier');
}
/**
 * Check if a node is a MemberExpression.
 *
 * @param node The AST node to check
 * @returns True if the node is a MemberExpression
 */
function isMember(node) {
    return isType(node, 'MemberExpression');
}
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            domInsertion: 'Direct DOM insertion is forbidden!'
        },
        schema: []
    },
    create(context) {
        const FORBIDDEN_DOM_INSERTION = ['insertBefore', 'appendChild', 'replaceChild'], FORBIDDEN_DOM_JQUERY_INSERTION = [
            'after',
            'before',
            'insertAfter',
            'insertBefore',
            'append',
            'prepend',
            'appendTo',
            'prependTo'
        ];
        const FORBIDDEN_METHODS = new Set([...FORBIDDEN_DOM_INSERTION, ...FORBIDDEN_DOM_JQUERY_INSERTION]);
        // --------------------------------------------------------------------------
        // Helpers
        // --------------------------------------------------------------------------
        /**
         * Process DOM insertion calls and report violations.
         *
         * @param node The call expression node to process
         */
        function processDomInsertion(node) {
            const callee = node.callee;
            if (isMember(callee)) {
                if (isIdentifier(callee.property) && 'name' in callee.property) {
                    if (FORBIDDEN_METHODS.has(callee.property.name)) {
                        context.report({ node: node, messageId: 'domInsertion' });
                    }
                }
            }
        }
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'CallExpression': function (node) {
                processDomInsertion(node);
            }
        };
    }
};
exports.default = rule;
