// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-width-including-column-header.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const odata_annotation_core_1 = require("@sap-ux/odata-annotation-core");
const rule_factory_1 = require("../language/rule-factory");
const diagnostics_1 = require("../language/diagnostics");
const annotations_1 = require("../project-context/linker/annotations");
const rule_fixer_1 = require("../language/rule-fixer");
/**
 * Determines if a table should have the widthIncludingColumnHeader property set.
 * Checks if the table's LineItem annotation contains DataFieldForAction or DataFieldForIntentBasedNavigation.
 *
 * @param table - Table to check
 * @param aliasInfo - Alias information for resolving qualified names
 * @returns True if the table should have widthIncludingColumnHeader set
 */
function shouldTableHaveWidthIncludingColumnHeader(table, aliasInfo) {
    if (!table.annotation) {
        return false;
    }
    const [collection] = (0, odata_annotation_core_1.elementsWithName)("Collection" /* Edm.Collection */, table.annotation.annotation.top.value);
    if (!collection) {
        return false;
    }
    const records = (0, odata_annotation_core_1.elements)((element) => {
        if (element.name !== "Record" /* Edm.Record */) {
            return false;
        }
        return (0, annotations_1.getRecordType)(aliasInfo, element) === 'com.sap.vocabularies.UI.v1.DataField';
    }, collection);
    return (records.length < 6 && records.length > 0 && table.configuration.widthIncludingColumnHeader.valueInFile !== true);
}
/**
 * Checks tables in a page for widthIncludingColumnHeader configuration issues.
 * Adds diagnostic problems for tables that should have this property set.
 *
 * @param page - SAP Fiori elements for OData V4 page to check (object page or list report)
 * @param parsedApp - Parsed application containing manifest.json file data
 * @param parsedService - Parsed service containing metadata and annotations
 * @param problems - Array to collect diagnostic problems
 */
function checkTablesInPage(page, parsedApp, parsedService, problems) {
    for (const table of page.lookup['table'] ?? []) {
        if (!table.annotation) {
            continue;
        }
        const aliasInfo = parsedService.artifacts.aliasInfo[table.annotation.annotation.top.uri];
        if (shouldTableHaveWidthIncludingColumnHeader(table, aliasInfo)) {
            problems.push({
                type: diagnostics_1.WIDTH_INCLUDING_COLUMN_HEADER_RULE_TYPE,
                pageName: page.targetName,
                manifest: {
                    uri: parsedApp.manifest.manifestUri,
                    object: parsedApp.manifestObject,
                    propertyPath: table.configuration.widthIncludingColumnHeader.configurationPath
                },
                annotation: {
                    file: table.annotation.annotation.source,
                    annotationPath: table.annotation.annotationPath,
                    reference: table.annotation.annotation.top
                }
            });
        }
    }
}
const rule = (0, rule_factory_1.createFioriRule)({
    ruleId: diagnostics_1.WIDTH_INCLUDING_COLUMN_HEADER_RULE_TYPE,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'By default, the column width is calculated based on the type of the content. You can include the column header in the width calculation by setting this property to true',
            url: 'https://ui5.sap.com/#/topic/c0f6592a592e47f9bb6d09900de47412'
        },
        messages: {
            ['width-including-column-header-manifest']: 'Small tables (< 6 columns) should use widthIncludingColumnHeader: true for improved calculation of the column width. Add it to the control configuration for "{{table}}" table.',
            ['width-including-column-header']: 'Small tables (< 6 columns) should use widthIncludingColumnHeader: true for improved calculation of the column width.'
        },
        fixable: 'code'
    },
    check(context) {
        const problems = [];
        for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
            if (app.type !== 'fe-v4') {
                continue;
            }
            for (const page of app.pages) {
                const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
                const parsedService = context.sourceCode.projectContext.getIndexedServiceForMainService(parsedApp);
                if (!parsedService) {
                    continue;
                }
                checkTablesInPage(page, parsedApp, parsedService, problems);
            }
        }
        return problems;
    },
    createJsonVisitorHandler: (context, diagnostic, paths) => function report(node) {
        context.report({
            node,
            messageId: 'width-including-column-header-manifest',
            data: {
                table: diagnostic.annotation.annotationPath
            },
            fix: (0, rule_fixer_1.createJsonFixer)({ context, node, deepestPathResult: paths, value: true })
        });
    },
    createAnnotations(context, validationResult) {
        if (validationResult.length === 0) {
            return {};
        }
        const lookup = new Set();
        for (const diagnostic of validationResult) {
            lookup.add(diagnostic.annotation?.reference?.value);
        }
        return {
            ['target>element[name="Annotation"]'](node) {
                if (!lookup.has(node)) {
                    return;
                }
                context.report({
                    node: node,
                    messageId: 'width-including-column-header'
                });
            }
        };
    }
});
exports.default = rule;
