// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-ui5-global-eval.js
"use strict";
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
/**
 * Check if a node is of a specific type.
 *
 * @param node The AST node to check
 * @param type The type to check for
 * @returns True if the node is of the specified type
 */
function isType(node, type) {
    return node?.type === type;
}
/**
 * Check if an array contains a specific object.
 *
 * @param a The array to search in
 * @param obj The object to search for
 * @returns True if the array contains the object
 */
function contains(a, obj) {
    return a.includes(obj);
}
/**
 * Check if a node is an Identifier.
 *
 * @param node The AST node to check
 * @returns True if the node is an Identifier
 */
function isIdentifier(node) {
    return isType(node, 'Identifier');
}
/**
 * Check if a node is a MemberExpression.
 *
 * @param node The AST node to check
 * @returns True if the node is a MemberExpression
 */
function isMember(node) {
    return isType(node, 'MemberExpression');
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            globalEval: 'Usage of globalEval() / eval() is not allowed due to strict Content Security Policy.'
        },
        schema: []
    },
    create(context) {
        const INTERESTING_METHODS_JQUERY = ['globalEval'];
        // --------------------------------------------------------------------------
        // Basic Helpers
        // --------------------------------------------------------------------------
        /**
         * Check if a node represents an interesting eval-related call.
         *
         * @param node The AST node to analyze
         * @returns True if the node represents an interesting eval-related call
         */
        function isInteresting(node) {
            const callNode = node;
            const callee = callNode.callee;
            if (isMember(callee)) {
                const memberNode = callee;
                const identifierProp = memberNode.property;
                if (isIdentifier(memberNode.property) && contains(INTERESTING_METHODS_JQUERY, identifierProp.name)) {
                    return true;
                }
            }
            return false;
        }
        return {
            'CallExpression': function (node) {
                if (isInteresting(node)) {
                    context.report({ node: node, messageId: 'globalEval' });
                }
            }
        };
    }
};
exports.default = rule;
