// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-location-usage.js
"use strict";
/**
 * @file Detect some warning for usages of (window.)document APIs
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            locationOverride: 'Writing to location is not allowed. Please consider using sap.m.URLHelper instead.',
            locationAssign: 'Usage of location.assign()'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        const LOCATION_OBJECTS = [];
        // Initialize factory functions
        const isWindowObject = (0, helpers_1.createIsWindowObject)(WINDOW_OBJECTS);
        const isLocation = (0, helpers_1.createIsLocation)(isWindowObject);
        const isLocationObject = (0, helpers_1.createIsLocationObject)(LOCATION_OBJECTS, isLocation);
        const rememberLocation = (0, helpers_1.createRememberLocation)(LOCATION_OBJECTS, isLocationObject);
        // --------------------------------------------------------------------------
        // Helpers
        // --------------------------------------------------------------------------
        /**
         * Check assignment expressions against location override violations.
         *
         * @param node The assignment expression node to check
         */
        function checkAssignmentAgainstOverride(node) {
            const identifier = node.left;
            if (isLocation(identifier) || // location = * || window.location = *
                isLocationObject(identifier.object)) {
                // location.* = || window.location.* = || const l = location; l.* =
                context.report({ node: node, messageId: 'locationOverride' });
            }
        }
        /**
         * Process member expressions for location.assign violations.
         *
         * @param node The member expression node to process
         */
        function processMemberExpression(node) {
            if (isLocationObject(node.object) && node.property.name === 'assign') {
                context.report({ node: node, messageId: 'locationAssign' });
            }
        }
        /**
         * Remember window object assignments for tracking references.
         *
         * @param left The left-hand side of the assignment
         * @param right The right-hand side of the assignment
         * @returns True if window object was remembered, false otherwise
         */
        function rememberWindow(left, right) {
            if (isWindowObject(right) && (0, helpers_1.isIdentifier)(left)) {
                WINDOW_OBJECTS.push(left.name);
                return true;
            }
            return false;
        }
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator': function (node) {
                return rememberWindow(node.id, node.init) || rememberLocation(node.id, node.init);
            },
            'AssignmentExpression': checkAssignmentAgainstOverride,
            'MemberExpression': processMemberExpression
        };
    }
};
exports.default = rule;
