// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-global-variable.js
"use strict";
/**
 * @file flag global variable declaration
 */
Object.defineProperty(exports, "__esModule", { value: true });
//------------------------------------------------------------------------------
// Helper Functions
//------------------------------------------------------------------------------
/**
 * Check if an array contains a specific item.
 *
 * @param array The array to search in
 * @param item The item to search for
 * @returns True if the array contains the item, false otherwise
 */
function contains(array, item) {
    return array.includes(item);
}
//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'disallow global variable declarations',
            category: 'Best Practices',
            recommended: true
        },
        fixable: undefined,
        schema: [],
        messages: {
            globalVariableNotAllowed: "Global variable '{{name}}' is not allowed"
        }
    },
    create(context) {
        const ALLOWED_VARIABLES = ['undefined', 'NaN', 'arguments', 'PDFJS', 'console', 'Infinity'];
        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------
        return {
            VariableDeclaration(node) {
                const sourceCode = context.sourceCode ?? context.getSourceCode();
                const scope = sourceCode.getScope
                    ? sourceCode.getScope(node)
                    : context.getScope();
                // Check if this is a global/module scope variable declaration
                if (scope.type === 'global' || scope.type === 'module') {
                    node.declarations.forEach((declaration) => {
                        if (declaration.id?.type === 'Identifier') {
                            const name = declaration.id.name;
                            if (!contains(ALLOWED_VARIABLES, name)) {
                                context.report({
                                    node: declaration.id,
                                    messageId: 'globalVariableNotAllowed',
                                    data: { name }
                                });
                            }
                        }
                    });
                }
            }
        };
    }
};
exports.default = rule;
