// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-global-selection.js
"use strict";
/**
 * @file Detect some warning for usages of (window.)document APIs
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            globalSelection: 'Global selection modification, only modify local selections'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        const FORBIDDEN_METHODS = ['getSelection'];
        // Create helper functions using the utility functions
        const isWindowObject = (0, helpers_1.createIsWindowObject)(WINDOW_OBJECTS);
        const rememberWindow = (0, helpers_1.createRememberWindow)(WINDOW_OBJECTS, isWindowObject);
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator'(node) {
                return rememberWindow(node.id, node.init);
            },
            'AssignmentExpression'(node) {
                return rememberWindow(node.left, node.right);
            },
            'MemberExpression'(node) {
                if (node &&
                    isWindowObject(node.object) &&
                    (0, helpers_1.isIdentifier)(node.property) &&
                    'name' in node.property &&
                    (0, helpers_1.contains)(FORBIDDEN_METHODS, node.property.name)) {
                    context.report({ node: node, messageId: 'globalSelection' });
                }
            }
        };
    }
};
exports.default = rule;
