// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-table-column-vertical-alignment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const diagnostics_1 = require("../language/diagnostics");
const rule_factory_1 = require("../language/rule-factory");
const rule_fixer_1 = require("../language/rule-fixer");
const version_1 = require("../utils/version");
const rule = (0, rule_factory_1.createFioriRule)({
    ruleId: diagnostics_1.TABLE_COLUMN_VERTICAL_ALIGNMENT,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'Sets a vertical alignment of columns in Responsive type tables',
            url: 'https://github.com/SAP/open-ux-tools/blob/main/packages/eslint-plugin-fiori-tools/docs/rules/sap-table-column-vertical-alignment.md'
        },
        messages: {
            [diagnostics_1.TABLE_COLUMN_VERTICAL_ALIGNMENT]: '"tableColumnVerticalAlignment" for Responsive tables must be set to "Middle" - the default value'
        },
        fixable: 'code'
    },
    check(context) {
        for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
            if (app.type !== 'fe-v2') {
                continue;
            }
            const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
            if (!parsedApp.manifest.minUI5Version ||
                (0, version_1.isLowerThanMinimalUi5Version)(parsedApp.manifest.minUI5Version, { major: 1, minor: 75 })) {
                continue;
            }
            const columnVerticalAlignment = app.configuration.tableColumnVerticalAlignment?.valueInFile;
            if (!columnVerticalAlignment || columnVerticalAlignment === 'Middle') {
                continue;
            }
            const responsiveTable = app.pages.find((page) => (page.lookup['table'] ?? []).find((table) => table.configuration.tableType.valueInFile === 'ResponsiveTable'));
            if (!responsiveTable) {
                continue;
            }
            return [
                {
                    type: diagnostics_1.TABLE_COLUMN_VERTICAL_ALIGNMENT,
                    manifest: {
                        uri: parsedApp.manifest.manifestUri,
                        object: parsedApp.manifestObject,
                        propertyPath: app.configuration.tableColumnVerticalAlignment.configurationPath
                    }
                }
            ];
        }
        return [];
    },
    createJsonVisitorHandler: (context, diagnostic, deepestPathResult) => function report(node) {
        context.report({
            node,
            messageId: diagnostics_1.TABLE_COLUMN_VERTICAL_ALIGNMENT,
            fix: (0, rule_fixer_1.createJsonFixer)({
                context,
                deepestPathResult,
                node,
                operation: 'delete'
            })
        });
    }
});
exports.default = rule;
