// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-hardcoded-url.js
"use strict";
/**
 * @file Rule to flag use of a hardcoded URL
 */
Object.defineProperty(exports, "__esModule", { value: true });
// ------------------------------------------------------------------------------
// Helper Functions
// ------------------------------------------------------------------------------
/**
 * RegExp pattern for hardcoded URL detection.
 */
const HARDCODED_URL_PATTERN = /(http|https):\/\/.*/;
/**
 * Check if a string contains prohibited hardcoded URLs.
 *
 * @param name The string to check for URLs
 * @returns RegExp exec result if URLs found, null otherwise
 */
function matchProhibited(name) {
    return HARDCODED_URL_PATTERN.exec(name);
}
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            hardcodedUrl: 'Hardcoded (non relative) URL found.'
        },
        schema: []
    },
    create(context) {
        const ALLOWED_DOMAINS = [
            // http://www.w3.org/2000/svg
            'http://www.w3.org/',
            // While this is not a true domain, adding the 'http version' here
            // is way easier and safer than trying to modify the regexp.
            'HTTP/1.1',
            'https://www.sap.com/Protocols/',
            'https://www.sap.com/adt',
            // localhost
            'http://localhost/offline/',
            'https://localhost/offline/'
        ];
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'Literal': function (node) {
                const literalNode = node;
                const val = literalNode.value;
                let result;
                if (typeof val === 'string') {
                    result = matchProhibited(val);
                    if (result && !ALLOWED_DOMAINS.some((domain) => val.includes(domain))) {
                        context.report({ node: node, messageId: 'hardcodedUrl' });
                    }
                }
            }
        };
    }
};
exports.default = rule;
