// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-creation-mode-for-table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const rule_factory_1 = require("../language/rule-factory");
const diagnostics_1 = require("../language/diagnostics");
const rule_fixer_1 = require("../language/rule-fixer");
const RECOMMENDED_MODE_V2 = 'creationRows';
const RECOMMENDED_MODE_V4_RESPONSIVE_GRID = 'InlineCreationRows';
const RECOMMENDED_MODE_V4_TREE = 'Inline';
/**
 * Reports a diagnostic issue related to the creation mode configuration.
 *
 * @param problems - Array to collect diagnostic problems
 * @param options - Diagnostic options
 * @param options.messageId - The message identifier for the diagnostic
 * @param options.pageName - Name of the page where the issue occurs
 * @param options.parsedApp - Parsed application context
 * @param options.configurationPath - Path to the configuration in manifest
 * @param options.tableType - Type of table (for example, 'GridTable' or 'ResponsiveTable')
 * @param options.validValues - Valid values for the configuration
 * @param options.recommendedValue - Recommended value for the configuration
 */
function reportDiagnostic(problems, { messageId, pageName, parsedApp, configurationPath, tableType, validValues = [], recommendedValue }) {
    problems.push({
        type: diagnostics_1.CREATION_MODE_FOR_TABLE,
        messageId,
        pageName,
        tableType,
        validValues,
        recommendedValue,
        manifest: {
            uri: parsedApp.manifest.manifestUri,
            object: parsedApp.manifestObject,
            propertyPath: configurationPath
        }
    });
}
/**
 * Checks if an analytical table has createMode configured at any level in OData V2 applications.
 * Analytical tables do not support creation mode and this function reports diagnostics, if configured.
 *
 * @param tableType - Type of the table
 * @param sectionCreateMode - Create mode configuration at the section level
 * @param pageCreateMode - Create mode configuration at the page level
 * @param appCreateMode - Create mode configuration at the application level
 * @param pageName - Name of the page
 * @param parsedApp - Parsed application context
 * @param problems - Array to collect diagnostic problems
 * @returns True, if an analytical table issue was found and reported. Otherwise, false
 */
function checkAnalyticalTableV2(tableType, sectionCreateMode, pageCreateMode, appCreateMode, pageName, parsedApp, problems) {
    if (tableType !== 'AnalyticalTable') {
        return false;
    }
    if (sectionCreateMode.valueInFile) {
        reportDiagnostic(problems, {
            messageId: 'analyticalTableNotSupported',
            pageName,
            parsedApp,
            configurationPath: sectionCreateMode.configurationPath,
            tableType
        });
        return true;
    }
    if (pageCreateMode.valueInFile) {
        reportDiagnostic(problems, {
            messageId: 'analyticalTableNotSupported',
            pageName,
            parsedApp,
            configurationPath: pageCreateMode.configurationPath,
            tableType
        });
        return true;
    }
    // table type is AnalyticalTable, but there is not creation mode configured - is valid.
    return false;
}
/**
 * Validates the create mode configuration for OData V2 applications.
 * Checks if the value is valid and recommends a value according to best practice.
 *
 * @param createMode - Create mode configuration to validate
 * @param pageName - Name of the page
 * @param parsedApp - Parsed application context
 * @param tableType - Type of the table
 * @param problems - Array to collect diagnostic problems
 * @returns True, if a configuration is found and an issue reported. False, if no configuration exists
 */
function validateCreateModeV2(createMode, pageName, parsedApp, tableType, problems) {
    if (!createMode.valueInFile) {
        return false;
    }
    if (!createMode.values.includes(createMode.valueInFile)) {
        reportDiagnostic(problems, {
            messageId: 'invalidCreateMode',
            pageName,
            parsedApp,
            configurationPath: createMode.configurationPath,
            tableType,
            validValues: createMode.values,
            recommendedValue: RECOMMENDED_MODE_V2
        });
        return true;
    }
    if (createMode.valueInFile !== RECOMMENDED_MODE_V2) {
        reportDiagnostic(problems, {
            messageId: 'recommendCreationRows',
            pageName,
            parsedApp,
            configurationPath: createMode.configurationPath,
            tableType,
            recommendedValue: RECOMMENDED_MODE_V2
        });
    }
    return true;
}
/**
 * Processes a single table in a OData V2 application.
 * Validates create mode at the section, page, and application levels with the correct priority.
 *
 * @param table - The table node to process
 * @param page - The object page containing the table
 * @param appCreateMode - Application-level create mode configuration
 * @param parsedApp - Parsed application context
 * @param problems - Array to collect diagnostic problems
 */
function processTableV2(table, page, appCreateMode, parsedApp, problems) {
    const sectionCreateMode = table.configuration.createMode;
    const pageCreateMode = page.configuration.createMode;
    const tableType = table.configuration.tableType?.valueInFile ?? '';
    // Check analytical table
    if (checkAnalyticalTableV2(tableType, sectionCreateMode, pageCreateMode, appCreateMode, page.targetName, parsedApp, problems)) {
        return;
    }
    // Check section, page, then app level
    if (validateCreateModeV2(sectionCreateMode, page.targetName, parsedApp, tableType, problems)) {
        return;
    }
    if (validateCreateModeV2(pageCreateMode, page.targetName, parsedApp, tableType, problems)) {
        return;
    }
    if (validateCreateModeV2(appCreateMode, page.targetName, parsedApp, tableType, problems)) {
        return;
    }
    // Suggest app level only once
    if (!problems.some((p) => p.messageId === 'suggestAppLevel' && p.manifest.uri === parsedApp.manifest.manifestUri)) {
        reportDiagnostic(problems, {
            messageId: 'suggestAppLevel',
            pageName: '',
            parsedApp,
            configurationPath: appCreateMode.configurationPath,
            tableType,
            recommendedValue: RECOMMENDED_MODE_V2
        });
    }
}
/**
 * Checks if an analytical table has creationMode configured at any level in OData V4 applications.
 * Analytical tables do not support creation mode and this function reports diagnostics, if configured.
 *
 * @param tableType - Type of the table
 * @param tableCreationMode - Creation mode configuration at the table level
 * @param appCreateMode - Create mode configuration at the application level
 * @param pageName - Name of the page
 * @param parsedApp - Parsed application context
 * @param problems - Array to collect diagnostic problems
 * @returns True, if an analytical table issue was found and reported. Otherwise, false
 */
function checkAnalyticalTableV4(tableType, tableCreationMode, appCreateMode, pageName, parsedApp, problems) {
    if (tableType !== 'AnalyticalTable') {
        return false;
    }
    if (tableCreationMode.valueInFile) {
        // Remove 'name' segment to delete the entire creationMode object
        const pathWithoutName = tableCreationMode.configurationPath.slice(0, -1);
        reportDiagnostic(problems, {
            messageId: 'analyticalTableNotSupported',
            pageName,
            parsedApp,
            configurationPath: pathWithoutName,
            tableType
        });
        return true;
    }
    // table type is AnalyticalTable, but there is not creation mode configured - is valid.
    return true;
}
/**
 * Gets the recommended creation mode value based on the table type for OData V4 applications.
 *
 * @param tableType - Type of the table
 * @returns 'Inline' for TreeTable and 'InlineCreationRows' for other table types
 */
function getRecommendedValueV4(tableType) {
    return tableType === 'TreeTable' ? RECOMMENDED_MODE_V4_TREE : RECOMMENDED_MODE_V4_RESPONSIVE_GRID;
}
/**
 * Validates the creation mode configuration for OData V4 applications.
 * Checks if the value is valid and recommends a value based on best practices and the table type.
 *
 * @param creationMode - Creation mode configuration to validate
 * @param pageName - Name of the page
 * @param parsedApp - Parsed application context
 * @param tableType - Type of the table
 * @param recommendedValue - The recommended value for this table type
 * @param problems - Array to collect diagnostic problems
 * @param shouldSuggestAppLevel - Whether to report on page level or not based on app level suggestion
 * @returns True, if a configuration was found and issue was reported. False, if no configuration exists
 */
function validateCreationModeV4(creationMode, pageName, parsedApp, tableType, recommendedValue, problems, shouldSuggestAppLevel) {
    const value = creationMode.valueInFile;
    if (value === undefined && shouldSuggestAppLevel === true) {
        return false;
    }
    const validValues = creationMode.values;
    // If shouldSuggestAppLevel is false, we MUST report on page level. (missing configuration)
    if (!value) {
        reportDiagnostic(problems, {
            messageId: 'invalidCreateModeV4',
            pageName,
            parsedApp,
            configurationPath: creationMode.configurationPath,
            tableType,
            validValues,
            recommendedValue
        });
        return true;
    }
    if (!validValues.includes(value)) {
        reportDiagnostic(problems, {
            messageId: 'invalidCreateModeV4',
            pageName,
            parsedApp,
            configurationPath: creationMode.configurationPath,
            tableType,
            validValues,
            recommendedValue
        });
        return true;
    }
    if (value !== recommendedValue) {
        reportDiagnostic(problems, {
            messageId: 'recommendInlineCreationRowsV4',
            pageName,
            parsedApp,
            configurationPath: creationMode.configurationPath,
            tableType,
            validValues,
            recommendedValue
        });
    }
    return true;
}
/**
 * Processes a single table in a OData V4 application.
 * Validates creation mode at the table and application levels with the correct priority.
 * Different table types have different recommended values.
 *
 * @param table - The table node to process
 * @param page - The object page containing the table
 * @param appCreateMode - Application-level create mode configuration
 * @param parsedApp - Parsed application context
 * @param problems - Array to collect diagnostic problems
 * @param shouldSuggestAppLevel - Whether to suggest app-level configuration (only if all tables have same recommended value)
 */
function processTableV4(table, page, appCreateMode, parsedApp, problems, shouldSuggestAppLevel) {
    const tableCreationMode = table.configuration.creationMode;
    const tableType = table.configuration.tableType?.valueInFile ?? '';
    // Check analytical table
    if (checkAnalyticalTableV4(tableType, tableCreationMode, appCreateMode, page.targetName, parsedApp, problems)) {
        return;
    }
    const recommendedValue = getRecommendedValueV4(tableType);
    // Check table level
    if (validateCreationModeV4(tableCreationMode, page.targetName, parsedApp, tableType, recommendedValue, problems, shouldSuggestAppLevel)) {
        return;
    }
    // Check app level
    if (appCreateMode.valueInFile) {
        const validValues = appCreateMode.values;
        if (!validValues.includes(appCreateMode.valueInFile)) {
            reportDiagnostic(problems, {
                messageId: 'invalidCreateModeV4',
                pageName: page.targetName,
                parsedApp,
                configurationPath: appCreateMode.configurationPath,
                tableType,
                validValues,
                recommendedValue
            });
            return;
        }
        if (appCreateMode.valueInFile !== recommendedValue) {
            reportDiagnostic(problems, {
                messageId: 'recommendInlineCreationRowsV4',
                pageName: page.targetName,
                parsedApp,
                configurationPath: appCreateMode.configurationPath,
                tableType,
                validValues,
                recommendedValue
            });
        }
        return;
    }
    // Suggest app level only once and only if all tables have the same recommended value
    if (shouldSuggestAppLevel &&
        !problems.some((p) => p.messageId === 'suggestAppLevelV4' && p.manifest.uri === parsedApp.manifest.manifestUri)) {
        reportDiagnostic(problems, {
            messageId: 'suggestAppLevelV4',
            pageName: '',
            parsedApp,
            configurationPath: appCreateMode.configurationPath,
            tableType,
            recommendedValue
        });
    }
}
/**
 * Processes all OData V2 applications in the project context.
 * Iterates through apps, pages, and tables to validate the creation mode configuration.
 *
 * @param context - Rule context containing source code and project information
 * @param problems - Array to collect diagnostic problems
 */
function processV2Apps(context, problems) {
    for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
        if (app.type !== 'fe-v2') {
            continue;
        }
        const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
        const appCreateMode = app.configuration.createMode;
        for (const page of app.pages) {
            if (page.type !== 'object-page') {
                continue;
            }
            for (const table of page.lookup['table'] ?? []) {
                processTableV2(table, page, appCreateMode, parsedApp, problems);
            }
        }
    }
}
/**
 * Determines if app-level creation mode suggestion is appropriate for V4 applications.
 * App-level suggestion is only made when all tables have the same recommended value,
 * or when there are only ResponsiveTable and GridTable types (which share the same recommended value).
 *
 * @param app - The linked application to analyze
 * @returns True if app-level suggestion should be made, false otherwise
 */
function shouldSuggestAppLevelV4(app) {
    // Collect all table types and their recommended values to determine if app-level suggestion is appropriate
    const recommendedValues = new Set();
    for (const page of app.pages) {
        if (page.type !== 'object-page') {
            continue;
        }
        for (const table of page.lookup['table'] ?? []) {
            const tableType = table.configuration.tableType?.valueInFile ?? '';
            if (tableType !== 'AnalyticalTable') {
                recommendedValues.add(getRecommendedValueV4(tableType));
            }
        }
    }
    // Only suggest app level if all tables have the same recommended value or if the size is two and values are 'ResponsiveTable' and 'GridTable'
    return (recommendedValues.size === 1 ||
        (recommendedValues.size === 2 && recommendedValues.has('ResponsiveTable') && recommendedValues.has('GridTable')));
}
/**
 * Processes all V4 applications in the project context.
 * Iterates through apps, pages, and tables to validate creation mode configuration.
 * Processes all OData V4 applications in the project context.
 * Iterates through apps, pages, and tables to validate the creation mode configuration.
 *
 * @param context - Rule context containing source code and project information
 * @param problems - Array to collect diagnostic problems
 */
function processV4Apps(context, problems) {
    for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
        if (app.type !== 'fe-v4') {
            continue;
        }
        const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
        const appCreateMode = app.configuration.createMode;
        const shouldSuggestAppLevel = shouldSuggestAppLevelV4(app);
        for (const page of app.pages) {
            if (page.type !== 'object-page') {
                continue;
            }
            for (const table of page.lookup['table'] ?? []) {
                processTableV4(table, page, appCreateMode, parsedApp, problems, shouldSuggestAppLevel);
            }
        }
    }
}
const rule = (0, rule_factory_1.createFioriRule)({
    ruleId: diagnostics_1.CREATION_MODE_FOR_TABLE,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'Ensure proper creationMode configuration for tables in SAP Fiori elements V2 and V4 applications',
            url: 'https://github.com/SAP/open-ux-tools/blob/main/packages/eslint-plugin-fiori-tools/docs/rules/sap-creation-mode-for-table.md'
        },
        messages: {
            invalidCreateMode: 'Invalid createMode value: "{{value}}"{{tableType}}. The recommended value is "creationRows".{{validValues}}.',
            recommendCreationRows: 'Consider using "creationRows" for a better user experience instead of "{{value}}".',
            suggestAppLevel: 'Consider adding createMode at the application level for a better user experience.',
            analyticalTableNotSupported: 'Creation mode is not supported for analytical tables. Remove the createMode or creationMode property.',
            invalidCreateModeV4: 'Invalid creationMode value "{{value}}"{{tableType}}. The recommended value is "{{recommendedValue}}".{{validValues}}',
            recommendInlineCreationRowsV4: 'Consider using "{{recommendedValue}}" for a better user experience instead of "{{value}}".',
            suggestAppLevelV4: 'Consider adding creationMode at the application level for better user experience.'
        },
        fixable: 'code'
    },
    check(context) {
        const problems = [];
        processV2Apps(context, problems);
        processV4Apps(context, problems);
        return problems;
    },
    createJsonVisitorHandler: (context, diagnostic, paths) => function report(node) {
        let tableType = '';
        if (diagnostic.tableType) {
            if (diagnostic.tableType === 'TreeTable') {
                tableType = ' for Tree Table';
            }
            else if (diagnostic.tableType === 'GridTable') {
                tableType = ' for Grid Table';
            }
            else {
                tableType = ` for ${diagnostic.tableType}`;
            }
        }
        let value = String(node.value);
        if (node.value.type === 'String') {
            value = node.value.value;
        }
        else if (node.name.type === 'String') {
            value = node.name.value;
        }
        const operation = diagnostic.messageId === 'analyticalTableNotSupported' ? 'delete' : undefined;
        context.report({
            node,
            messageId: diagnostic.messageId,
            data: {
                value,
                tableType,
                validValues: diagnostic.validValues.length > 0
                    ? ` Valid values are: ${diagnostic.validValues.join(', ')}.`
                    : '',
                recommendedValue: diagnostic.recommendedValue ?? ''
            },
            fix: (0, rule_fixer_1.createJsonFixer)({
                value: operation === 'delete' ? undefined : diagnostic.recommendedValue,
                context,
                deepestPathResult: paths,
                node,
                operation
            })
        });
    }
});
exports.default = rule;
