// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-anchor-bar-visible.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const diagnostics_1 = require("../language/diagnostics");
const rule_factory_1 = require("../language/rule-factory");
const rule_fixer_1 = require("../language/rule-fixer");
const rule = (0, rule_factory_1.createFioriRule)({
    ruleId: diagnostics_1.ANCHOR_BAR_VISIBLE,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'Anchor Bar Visible should not be set to false in manifest settings for object page headers (except form entry object pages)',
            url: 'https://github.com/SAP/open-ux-tools/blob/main/packages/eslint-plugin-fiori-tools/docs/rules/sap-anchor-bar-visible.md'
        },
        messages: {
            [diagnostics_1.ANCHOR_BAR_VISIBLE]: 'The "anchorBarVisible" property should not be set to false in manifest settings. Remove this property from the object page header configuration. Exception: Form entry object pages can have both "visible" and "anchorBarVisible" set to false.'
        },
        fixable: 'code'
    },
    check(context) {
        const problems = [];
        for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
            if (app.type !== 'fe-v4') {
                continue;
            }
            for (const page of app.pages) {
                if (page.type !== 'object-page') {
                    continue;
                }
                const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
                // Check if anchorBarVisible is set to false
                // Exception: Form Entry Object Pages can have both visible: false and anchorBarVisible: false
                const anchorBarVisible = page.header.anchorBarVisible.valueInFile;
                const headerVisible = page.header.visible.valueInFile;
                if (anchorBarVisible === false && headerVisible !== false) {
                    problems.push({
                        type: diagnostics_1.ANCHOR_BAR_VISIBLE,
                        pageName: page.targetName,
                        manifest: {
                            uri: parsedApp.manifest.manifestUri,
                            object: parsedApp.manifestObject,
                            propertyPath: page.header.anchorBarVisible.configurationPath
                        }
                    });
                }
            }
        }
        return problems;
    },
    createJsonVisitorHandler: (context, diagnostic, deepestPathResult) => function report(node) {
        context.report({
            node,
            messageId: diagnostics_1.ANCHOR_BAR_VISIBLE,
            fix: (0, rule_fixer_1.createJsonFixer)({
                context,
                deepestPathResult,
                node,
                operation: 'delete'
            })
        });
    }
});
exports.default = rule;
