// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-encode-file-service.js
"use strict";
/**
 * @file        Check "sap-no-encode-file-service" should detect the usage of "/sap/bc/ui2/encode_file".
 */
Object.defineProperty(exports, "__esModule", { value: true });
// ------------------------------------------------------------------------------
// Helper Functions
// ------------------------------------------------------------------------------
/**
 * Check if a value is a string.
 *
 * @param string The value to check
 * @returns True if the value is a string
 */
function isString(string) {
    return typeof string === 'string';
}
/**
 * Check if a string starts with a substring.
 *
 * @param base The base string to check
 * @param sub The substring to look for at the start
 * @returns True if the base string starts with the substring
 */
function startsWith(base, sub) {
    return base.startsWith(sub);
}
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            encodeFileService: "Usage of phrase '/sap/bc/ui2/encode_file' detected"
        },
        schema: []
    },
    create(context) {
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'Literal': function (node) {
                const value = node.value;
                if (isString(value) && startsWith(value, '/sap/bc/ui2/encode_file')) {
                    context.report({ node: node, messageId: 'encodeFileService' });
                }
            }
        };
    }
};
exports.default = rule;
