// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-element-creation.js
"use strict";
/**
 * @file Detect direct DOM insertion
 */
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../utils/helpers");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            category: 'Best Practices',
            recommended: false
        },
        messages: {
            elementCreation: 'Direct element creation, create a custom control instead'
        },
        schema: []
    },
    create(context) {
        const FORBIDDEN_DOM_INSERTION = [
            'createElement',
            'createTextNode',
            'createElementNS',
            'createDocumentFragment',
            'createComment',
            'createAttribute',
            'createEvent'
        ];
        const createVisitors = (0, helpers_1.createDocumentBasedRuleVisitors)({
            isInteresting: (node, isDocumentObject) => {
                return node && (0, helpers_1.isCall)(node.parent) && isDocumentObject(node.object);
            },
            isValid: (node) => {
                let methodName = false;
                if ((0, helpers_1.isIdentifier)(node.property)) {
                    methodName = node.property.name;
                }
                else if ((0, helpers_1.isLiteral)(node.property)) {
                    methodName = node.property.value;
                }
                return (methodName &&
                    (!(0, helpers_1.contains)(FORBIDDEN_DOM_INSERTION, methodName) ||
                        (methodName === 'createElement' &&
                            (0, helpers_1.isCall)(node.parent) &&
                            node.parent.arguments &&
                            node.parent.arguments.length > 0 &&
                            (0, helpers_1.isLiteral)(node.parent.arguments[0]) &&
                            node.parent.arguments[0].value === 'a')));
            },
            messageId: 'elementCreation'
        });
        return createVisitors(context);
    }
};
exports.default = rule;
