// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-enable-export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const diagnostics_1 = require("../language/diagnostics");
const rule_factory_1 = require("../language/rule-factory");
const rule_fixer_1 = require("../language/rule-fixer");
const rule = (0, rule_factory_1.createFioriRule)({
    ruleId: diagnostics_1.ENABLE_EXPORT,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'Sets whether to enable or disable the "Export" button in tables',
            url: 'https://github.com/SAP/open-ux-tools/blob/main/packages/eslint-plugin-fiori-tools/docs/rules/sap-enable-export.md'
        },
        messages: {
            [diagnostics_1.ENABLE_EXPORT]: 'Export functionality in the table must be enabled'
        },
        fixable: 'code'
    },
    check(context) {
        const problems = [];
        for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
            const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
            const parsedService = context.sourceCode.projectContext.getIndexedServiceForMainService(parsedApp);
            if (!parsedService) {
                continue;
            }
            if (app.type === 'fe-v4') {
                for (const page of app.pages) {
                    problems.push(...handleExportInTableV4(page, parsedApp));
                }
            }
        }
        return problems;
    },
    createJsonVisitorHandler: (context, diagnostic, deepestPathResult) => function report(node) {
        context.report({
            node,
            messageId: diagnostics_1.ENABLE_EXPORT,
            fix: (0, rule_fixer_1.createJsonFixer)({
                context,
                deepestPathResult,
                node,
                operation: 'delete'
            })
        });
    }
});
/**
 * Looks through V4 app page tables and returns problems if enableExport is set to false.
 *
 * @param page - V4 app page
 * @param parsedApp - parsed V4 app
 * @returns - EnableExport issues
 */
function handleExportInTableV4(page, parsedApp) {
    const problems = [];
    for (const table of page.lookup['table'] ?? []) {
        if (table.configuration.enableExport.valueInFile === false) {
            problems.push({
                type: diagnostics_1.ENABLE_EXPORT,
                pageName: page.targetName,
                manifest: {
                    uri: parsedApp.manifest.manifestUri,
                    object: parsedApp.manifestObject,
                    propertyPath: table.configuration.enableExport.configurationPath
                }
            });
        }
    }
    return problems;
}
exports.default = rule;
